{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/GKD-stack/DSC180A-Methodology-1/blob/main/DSC180A__W8_9_ProjectPart1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "9cfa2da9",
      "metadata": {
        "id": "9cfa2da9"
      },
      "outputs": [],
      "source": [
        "import pandas_plink as pp\n",
        "import statsmodels.api as sm\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import glob"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "55388f50",
      "metadata": {
        "id": "55388f50"
      },
      "source": [
        "#### Take a look at the gene_expression, gene_annotation files we have, we will subset gene expression data to find genes that are protein coding"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0e649747",
      "metadata": {
        "id": "0e649747"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "bcb5475f",
      "metadata": {
        "id": "bcb5475f"
      },
      "outputs": [],
      "source": [
        "gene_expression_data = pd.read_csv('GD462.GeneQuantRPKM.50FN.samplename.resk10.txt', delimiter='\\t')\n",
        "\n",
        "#slice the string to make the gene symbols consistent with gene_annotation data\n",
        "gene_expression_data['TargetID'] = gene_expression_data['TargetID'].str.split('.').str[0]\n",
        "gene_expression_data['Gene_Symbol'] = gene_expression_data['Gene_Symbol'].str.split('.').str[0]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "726087b5",
      "metadata": {
        "id": "726087b5",
        "outputId": "c80a4696-c981-4cae-8be8-19831ef85a4a"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>TargetID</th>\n",
              "      <th>Gene_Symbol</th>\n",
              "      <th>Chr</th>\n",
              "      <th>Coord</th>\n",
              "      <th>HG00096</th>\n",
              "      <th>HG00097</th>\n",
              "      <th>HG00099</th>\n",
              "      <th>HG00100</th>\n",
              "      <th>HG00101</th>\n",
              "      <th>HG00102</th>\n",
              "      <th>...</th>\n",
              "      <th>NA20810</th>\n",
              "      <th>NA20811</th>\n",
              "      <th>NA20812</th>\n",
              "      <th>NA20813</th>\n",
              "      <th>NA20814</th>\n",
              "      <th>NA20815</th>\n",
              "      <th>NA20816</th>\n",
              "      <th>NA20819</th>\n",
              "      <th>NA20826</th>\n",
              "      <th>NA20828</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSG00000152931</td>\n",
              "      <td>ENSG00000152931</td>\n",
              "      <td>5</td>\n",
              "      <td>59783540</td>\n",
              "      <td>0.101858</td>\n",
              "      <td>0.078110</td>\n",
              "      <td>0.048981</td>\n",
              "      <td>0.118597</td>\n",
              "      <td>0.004035</td>\n",
              "      <td>0.010925</td>\n",
              "      <td>...</td>\n",
              "      <td>0.088601</td>\n",
              "      <td>0.240010</td>\n",
              "      <td>0.137175</td>\n",
              "      <td>0.148494</td>\n",
              "      <td>0.038643</td>\n",
              "      <td>0.088509</td>\n",
              "      <td>0.029204</td>\n",
              "      <td>0.024423</td>\n",
              "      <td>0.044816</td>\n",
              "      <td>0.139186</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000183696</td>\n",
              "      <td>ENSG00000183696</td>\n",
              "      <td>7</td>\n",
              "      <td>48128225</td>\n",
              "      <td>8.183805</td>\n",
              "      <td>5.686911</td>\n",
              "      <td>2.434653</td>\n",
              "      <td>3.830894</td>\n",
              "      <td>6.612288</td>\n",
              "      <td>4.709646</td>\n",
              "      <td>...</td>\n",
              "      <td>13.428205</td>\n",
              "      <td>6.094500</td>\n",
              "      <td>12.536000</td>\n",
              "      <td>2.217262</td>\n",
              "      <td>3.573394</td>\n",
              "      <td>7.583364</td>\n",
              "      <td>4.052882</td>\n",
              "      <td>1.570378</td>\n",
              "      <td>4.900372</td>\n",
              "      <td>6.737308</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000139269</td>\n",
              "      <td>ENSG00000139269</td>\n",
              "      <td>12</td>\n",
              "      <td>57846106</td>\n",
              "      <td>1.199910</td>\n",
              "      <td>1.573572</td>\n",
              "      <td>0.521616</td>\n",
              "      <td>1.447225</td>\n",
              "      <td>3.565791</td>\n",
              "      <td>1.982681</td>\n",
              "      <td>...</td>\n",
              "      <td>3.225880</td>\n",
              "      <td>1.996067</td>\n",
              "      <td>2.854923</td>\n",
              "      <td>2.267343</td>\n",
              "      <td>1.331201</td>\n",
              "      <td>2.187895</td>\n",
              "      <td>1.004250</td>\n",
              "      <td>3.003316</td>\n",
              "      <td>1.984362</td>\n",
              "      <td>1.684954</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000169129</td>\n",
              "      <td>ENSG00000169129</td>\n",
              "      <td>10</td>\n",
              "      <td>116164515</td>\n",
              "      <td>0.831940</td>\n",
              "      <td>0.069778</td>\n",
              "      <td>0.931086</td>\n",
              "      <td>0.620941</td>\n",
              "      <td>1.660668</td>\n",
              "      <td>0.570481</td>\n",
              "      <td>...</td>\n",
              "      <td>1.023381</td>\n",
              "      <td>1.127852</td>\n",
              "      <td>0.774409</td>\n",
              "      <td>1.495854</td>\n",
              "      <td>0.895342</td>\n",
              "      <td>1.513521</td>\n",
              "      <td>0.826377</td>\n",
              "      <td>1.021201</td>\n",
              "      <td>0.952502</td>\n",
              "      <td>0.740565</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSG00000134602</td>\n",
              "      <td>ENSG00000134602</td>\n",
              "      <td>X</td>\n",
              "      <td>131157293</td>\n",
              "      <td>27.646422</td>\n",
              "      <td>24.395572</td>\n",
              "      <td>16.445374</td>\n",
              "      <td>24.806650</td>\n",
              "      <td>25.113349</td>\n",
              "      <td>19.233988</td>\n",
              "      <td>...</td>\n",
              "      <td>25.079490</td>\n",
              "      <td>28.725528</td>\n",
              "      <td>24.450520</td>\n",
              "      <td>27.264069</td>\n",
              "      <td>26.912814</td>\n",
              "      <td>29.509210</td>\n",
              "      <td>26.462331</td>\n",
              "      <td>25.624009</td>\n",
              "      <td>25.707741</td>\n",
              "      <td>22.824957</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23717</th>\n",
              "      <td>ENSG00000235472</td>\n",
              "      <td>ENSG00000235472</td>\n",
              "      <td>13</td>\n",
              "      <td>29172970</td>\n",
              "      <td>31.582832</td>\n",
              "      <td>34.071123</td>\n",
              "      <td>19.394365</td>\n",
              "      <td>37.523721</td>\n",
              "      <td>33.430473</td>\n",
              "      <td>37.844966</td>\n",
              "      <td>...</td>\n",
              "      <td>47.905182</td>\n",
              "      <td>33.224977</td>\n",
              "      <td>39.827675</td>\n",
              "      <td>27.096811</td>\n",
              "      <td>34.686342</td>\n",
              "      <td>37.386766</td>\n",
              "      <td>30.598840</td>\n",
              "      <td>33.516674</td>\n",
              "      <td>32.017940</td>\n",
              "      <td>38.341888</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23718</th>\n",
              "      <td>ENSG00000114423</td>\n",
              "      <td>ENSG00000114423</td>\n",
              "      <td>3</td>\n",
              "      <td>105588396</td>\n",
              "      <td>14.054749</td>\n",
              "      <td>14.477899</td>\n",
              "      <td>11.584425</td>\n",
              "      <td>12.637956</td>\n",
              "      <td>12.015089</td>\n",
              "      <td>13.750655</td>\n",
              "      <td>...</td>\n",
              "      <td>11.723462</td>\n",
              "      <td>9.900372</td>\n",
              "      <td>10.473115</td>\n",
              "      <td>13.433413</td>\n",
              "      <td>15.832594</td>\n",
              "      <td>19.216176</td>\n",
              "      <td>10.213739</td>\n",
              "      <td>14.563192</td>\n",
              "      <td>15.637732</td>\n",
              "      <td>8.357117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23719</th>\n",
              "      <td>ENSG00000243312</td>\n",
              "      <td>ENSG00000243312</td>\n",
              "      <td>4</td>\n",
              "      <td>87791344</td>\n",
              "      <td>1.112114</td>\n",
              "      <td>0.831797</td>\n",
              "      <td>0.253228</td>\n",
              "      <td>0.271568</td>\n",
              "      <td>0.486086</td>\n",
              "      <td>1.362640</td>\n",
              "      <td>...</td>\n",
              "      <td>1.168991</td>\n",
              "      <td>0.645389</td>\n",
              "      <td>0.819469</td>\n",
              "      <td>0.515448</td>\n",
              "      <td>0.463054</td>\n",
              "      <td>1.580658</td>\n",
              "      <td>0.701396</td>\n",
              "      <td>0.771233</td>\n",
              "      <td>0.857330</td>\n",
              "      <td>0.703369</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23720</th>\n",
              "      <td>ENSG00000257337</td>\n",
              "      <td>ENSG00000257337</td>\n",
              "      <td>12</td>\n",
              "      <td>53448222</td>\n",
              "      <td>3.826396</td>\n",
              "      <td>6.045798</td>\n",
              "      <td>2.593872</td>\n",
              "      <td>4.447169</td>\n",
              "      <td>5.294657</td>\n",
              "      <td>4.106823</td>\n",
              "      <td>...</td>\n",
              "      <td>6.524398</td>\n",
              "      <td>6.017795</td>\n",
              "      <td>3.460273</td>\n",
              "      <td>3.905032</td>\n",
              "      <td>5.023161</td>\n",
              "      <td>5.333027</td>\n",
              "      <td>5.694370</td>\n",
              "      <td>8.142939</td>\n",
              "      <td>5.622043</td>\n",
              "      <td>6.026476</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23721</th>\n",
              "      <td>ENSG00000177494</td>\n",
              "      <td>ENSG00000177494</td>\n",
              "      <td>3</td>\n",
              "      <td>111314290</td>\n",
              "      <td>1.183398</td>\n",
              "      <td>0.892619</td>\n",
              "      <td>0.917497</td>\n",
              "      <td>3.504894</td>\n",
              "      <td>1.168247</td>\n",
              "      <td>3.110953</td>\n",
              "      <td>...</td>\n",
              "      <td>1.936493</td>\n",
              "      <td>1.177972</td>\n",
              "      <td>2.673129</td>\n",
              "      <td>13.892711</td>\n",
              "      <td>2.763749</td>\n",
              "      <td>3.032246</td>\n",
              "      <td>4.558199</td>\n",
              "      <td>2.807851</td>\n",
              "      <td>2.021329</td>\n",
              "      <td>3.181096</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>23722 rows × 466 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "              TargetID      Gene_Symbol Chr      Coord    HG00096    HG00097  \\\n",
              "0      ENSG00000152931  ENSG00000152931   5   59783540   0.101858   0.078110   \n",
              "1      ENSG00000183696  ENSG00000183696   7   48128225   8.183805   5.686911   \n",
              "2      ENSG00000139269  ENSG00000139269  12   57846106   1.199910   1.573572   \n",
              "3      ENSG00000169129  ENSG00000169129  10  116164515   0.831940   0.069778   \n",
              "4      ENSG00000134602  ENSG00000134602   X  131157293  27.646422  24.395572   \n",
              "...                ...              ...  ..        ...        ...        ...   \n",
              "23717  ENSG00000235472  ENSG00000235472  13   29172970  31.582832  34.071123   \n",
              "23718  ENSG00000114423  ENSG00000114423   3  105588396  14.054749  14.477899   \n",
              "23719  ENSG00000243312  ENSG00000243312   4   87791344   1.112114   0.831797   \n",
              "23720  ENSG00000257337  ENSG00000257337  12   53448222   3.826396   6.045798   \n",
              "23721  ENSG00000177494  ENSG00000177494   3  111314290   1.183398   0.892619   \n",
              "\n",
              "         HG00099    HG00100    HG00101    HG00102  ...    NA20810    NA20811  \\\n",
              "0       0.048981   0.118597   0.004035   0.010925  ...   0.088601   0.240010   \n",
              "1       2.434653   3.830894   6.612288   4.709646  ...  13.428205   6.094500   \n",
              "2       0.521616   1.447225   3.565791   1.982681  ...   3.225880   1.996067   \n",
              "3       0.931086   0.620941   1.660668   0.570481  ...   1.023381   1.127852   \n",
              "4      16.445374  24.806650  25.113349  19.233988  ...  25.079490  28.725528   \n",
              "...          ...        ...        ...        ...  ...        ...        ...   \n",
              "23717  19.394365  37.523721  33.430473  37.844966  ...  47.905182  33.224977   \n",
              "23718  11.584425  12.637956  12.015089  13.750655  ...  11.723462   9.900372   \n",
              "23719   0.253228   0.271568   0.486086   1.362640  ...   1.168991   0.645389   \n",
              "23720   2.593872   4.447169   5.294657   4.106823  ...   6.524398   6.017795   \n",
              "23721   0.917497   3.504894   1.168247   3.110953  ...   1.936493   1.177972   \n",
              "\n",
              "         NA20812    NA20813    NA20814    NA20815    NA20816    NA20819  \\\n",
              "0       0.137175   0.148494   0.038643   0.088509   0.029204   0.024423   \n",
              "1      12.536000   2.217262   3.573394   7.583364   4.052882   1.570378   \n",
              "2       2.854923   2.267343   1.331201   2.187895   1.004250   3.003316   \n",
              "3       0.774409   1.495854   0.895342   1.513521   0.826377   1.021201   \n",
              "4      24.450520  27.264069  26.912814  29.509210  26.462331  25.624009   \n",
              "...          ...        ...        ...        ...        ...        ...   \n",
              "23717  39.827675  27.096811  34.686342  37.386766  30.598840  33.516674   \n",
              "23718  10.473115  13.433413  15.832594  19.216176  10.213739  14.563192   \n",
              "23719   0.819469   0.515448   0.463054   1.580658   0.701396   0.771233   \n",
              "23720   3.460273   3.905032   5.023161   5.333027   5.694370   8.142939   \n",
              "23721   2.673129  13.892711   2.763749   3.032246   4.558199   2.807851   \n",
              "\n",
              "         NA20826    NA20828  \n",
              "0       0.044816   0.139186  \n",
              "1       4.900372   6.737308  \n",
              "2       1.984362   1.684954  \n",
              "3       0.952502   0.740565  \n",
              "4      25.707741  22.824957  \n",
              "...          ...        ...  \n",
              "23717  32.017940  38.341888  \n",
              "23718  15.637732   8.357117  \n",
              "23719   0.857330   0.703369  \n",
              "23720   5.622043   6.026476  \n",
              "23721   2.021329   3.181096  \n",
              "\n",
              "[23722 rows x 466 columns]"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "gene_expression_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "328317ea",
      "metadata": {
        "id": "328317ea"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "6402f9fe",
      "metadata": {
        "id": "6402f9fe"
      },
      "outputs": [],
      "source": [
        "gene_annotation_data = pd.read_csv('gene_annot.txt', delimiter='\\t')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "70a4b532",
      "metadata": {
        "id": "70a4b532",
        "outputId": "a6840128-da05-49fd-c5bb-a5630452354b"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>CHR</th>\n",
              "      <th>START</th>\n",
              "      <th>STOP</th>\n",
              "      <th>SYM</th>\n",
              "      <th>TYPE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>DDX11L1</td>\n",
              "      <td>1</td>\n",
              "      <td>11868</td>\n",
              "      <td>13052</td>\n",
              "      <td>ENSG00000223972</td>\n",
              "      <td>transcribed_unprocessed_pseudogene</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>OR4F5</td>\n",
              "      <td>1</td>\n",
              "      <td>69090</td>\n",
              "      <td>70008</td>\n",
              "      <td>ENSG00000186092</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FAM87B</td>\n",
              "      <td>1</td>\n",
              "      <td>817370</td>\n",
              "      <td>819834</td>\n",
              "      <td>ENSG00000177757</td>\n",
              "      <td>lincRNA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>LINC00115</td>\n",
              "      <td>1</td>\n",
              "      <td>826205</td>\n",
              "      <td>827522</td>\n",
              "      <td>ENSG00000225880</td>\n",
              "      <td>lincRNA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>LINC01128</td>\n",
              "      <td>1</td>\n",
              "      <td>827607</td>\n",
              "      <td>859446</td>\n",
              "      <td>ENSG00000228794</td>\n",
              "      <td>processed_transcript</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20495</th>\n",
              "      <td>ARSA</td>\n",
              "      <td>22</td>\n",
              "      <td>50622753</td>\n",
              "      <td>50628179</td>\n",
              "      <td>ENSG00000100299</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20496</th>\n",
              "      <td>SHANK3</td>\n",
              "      <td>22</td>\n",
              "      <td>50674414</td>\n",
              "      <td>50733298</td>\n",
              "      <td>ENSG00000251322</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20497</th>\n",
              "      <td>ACR</td>\n",
              "      <td>22</td>\n",
              "      <td>50738195</td>\n",
              "      <td>50745334</td>\n",
              "      <td>ENSG00000100312</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20498</th>\n",
              "      <td>RPL23AP82</td>\n",
              "      <td>22</td>\n",
              "      <td>50756947</td>\n",
              "      <td>50801309</td>\n",
              "      <td>ENSG00000184319</td>\n",
              "      <td>transcribed_unprocessed_pseudogene</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20499</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>22</td>\n",
              "      <td>50767500</td>\n",
              "      <td>50783663</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>20500 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "              ID  CHR     START      STOP              SYM  \\\n",
              "0        DDX11L1    1     11868     13052  ENSG00000223972   \n",
              "1          OR4F5    1     69090     70008  ENSG00000186092   \n",
              "2         FAM87B    1    817370    819834  ENSG00000177757   \n",
              "3      LINC00115    1    826205    827522  ENSG00000225880   \n",
              "4      LINC01128    1    827607    859446  ENSG00000228794   \n",
              "...          ...  ...       ...       ...              ...   \n",
              "20495       ARSA   22  50622753  50628179  ENSG00000100299   \n",
              "20496     SHANK3   22  50674414  50733298  ENSG00000251322   \n",
              "20497        ACR   22  50738195  50745334  ENSG00000100312   \n",
              "20498  RPL23AP82   22  50756947  50801309  ENSG00000184319   \n",
              "20499     RABL2B   22  50767500  50783663  ENSG00000079974   \n",
              "\n",
              "                                     TYPE  \n",
              "0      transcribed_unprocessed_pseudogene  \n",
              "1                          protein_coding  \n",
              "2                                 lincRNA  \n",
              "3                                 lincRNA  \n",
              "4                    processed_transcript  \n",
              "...                                   ...  \n",
              "20495                      protein_coding  \n",
              "20496                      protein_coding  \n",
              "20497                      protein_coding  \n",
              "20498  transcribed_unprocessed_pseudogene  \n",
              "20499                      protein_coding  \n",
              "\n",
              "[20500 rows x 6 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "gene_annotation_data"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "972c81bb",
      "metadata": {
        "id": "972c81bb"
      },
      "source": [
        "### Subset gene_expression data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "98506d05",
      "metadata": {
        "id": "98506d05"
      },
      "outputs": [],
      "source": [
        "# Filter for protein-coding genes\n",
        "protein_coding_genes = gene_annotation_data[gene_annotation_data['TYPE'] == 'protein_coding']\n",
        "\n",
        "# Get a list of protein-coding gene symbols\n",
        "protein_coding_symbols = protein_coding_genes['SYM'].tolist()\n",
        "\n",
        "# Subset gene_expression_data for protein-coding genes\n",
        "protein_coding_expression_data = gene_expression_data[gene_expression_data['Gene_Symbol'].isin(protein_coding_symbols)]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1bdf4a61",
      "metadata": {
        "id": "1bdf4a61",
        "outputId": "ece85162-073a-487e-9fb9-f5dbd70ebbea"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>TargetID</th>\n",
              "      <th>Gene_Symbol</th>\n",
              "      <th>Chr</th>\n",
              "      <th>Coord</th>\n",
              "      <th>HG00096</th>\n",
              "      <th>HG00097</th>\n",
              "      <th>HG00099</th>\n",
              "      <th>HG00100</th>\n",
              "      <th>HG00101</th>\n",
              "      <th>HG00102</th>\n",
              "      <th>...</th>\n",
              "      <th>NA20810</th>\n",
              "      <th>NA20811</th>\n",
              "      <th>NA20812</th>\n",
              "      <th>NA20813</th>\n",
              "      <th>NA20814</th>\n",
              "      <th>NA20815</th>\n",
              "      <th>NA20816</th>\n",
              "      <th>NA20819</th>\n",
              "      <th>NA20826</th>\n",
              "      <th>NA20828</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSG00000183696</td>\n",
              "      <td>ENSG00000183696</td>\n",
              "      <td>7</td>\n",
              "      <td>48128225</td>\n",
              "      <td>8.183805</td>\n",
              "      <td>5.686911</td>\n",
              "      <td>2.434653</td>\n",
              "      <td>3.830894</td>\n",
              "      <td>6.612288</td>\n",
              "      <td>4.709646</td>\n",
              "      <td>...</td>\n",
              "      <td>13.428205</td>\n",
              "      <td>6.094500</td>\n",
              "      <td>12.536000</td>\n",
              "      <td>2.217262</td>\n",
              "      <td>3.573394</td>\n",
              "      <td>7.583364</td>\n",
              "      <td>4.052882</td>\n",
              "      <td>1.570378</td>\n",
              "      <td>4.900372</td>\n",
              "      <td>6.737308</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSG00000139269</td>\n",
              "      <td>ENSG00000139269</td>\n",
              "      <td>12</td>\n",
              "      <td>57846106</td>\n",
              "      <td>1.199910</td>\n",
              "      <td>1.573572</td>\n",
              "      <td>0.521616</td>\n",
              "      <td>1.447225</td>\n",
              "      <td>3.565791</td>\n",
              "      <td>1.982681</td>\n",
              "      <td>...</td>\n",
              "      <td>3.225880</td>\n",
              "      <td>1.996067</td>\n",
              "      <td>2.854923</td>\n",
              "      <td>2.267343</td>\n",
              "      <td>1.331201</td>\n",
              "      <td>2.187895</td>\n",
              "      <td>1.004250</td>\n",
              "      <td>3.003316</td>\n",
              "      <td>1.984362</td>\n",
              "      <td>1.684954</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSG00000169129</td>\n",
              "      <td>ENSG00000169129</td>\n",
              "      <td>10</td>\n",
              "      <td>116164515</td>\n",
              "      <td>0.831940</td>\n",
              "      <td>0.069778</td>\n",
              "      <td>0.931086</td>\n",
              "      <td>0.620941</td>\n",
              "      <td>1.660668</td>\n",
              "      <td>0.570481</td>\n",
              "      <td>...</td>\n",
              "      <td>1.023381</td>\n",
              "      <td>1.127852</td>\n",
              "      <td>0.774409</td>\n",
              "      <td>1.495854</td>\n",
              "      <td>0.895342</td>\n",
              "      <td>1.513521</td>\n",
              "      <td>0.826377</td>\n",
              "      <td>1.021201</td>\n",
              "      <td>0.952502</td>\n",
              "      <td>0.740565</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>ENSG00000136237</td>\n",
              "      <td>ENSG00000136237</td>\n",
              "      <td>7</td>\n",
              "      <td>22396763</td>\n",
              "      <td>3.788503</td>\n",
              "      <td>2.050963</td>\n",
              "      <td>4.000313</td>\n",
              "      <td>3.271619</td>\n",
              "      <td>1.798216</td>\n",
              "      <td>1.516688</td>\n",
              "      <td>...</td>\n",
              "      <td>2.909393</td>\n",
              "      <td>1.921176</td>\n",
              "      <td>5.083873</td>\n",
              "      <td>2.866573</td>\n",
              "      <td>1.297788</td>\n",
              "      <td>2.888316</td>\n",
              "      <td>2.145022</td>\n",
              "      <td>3.557598</td>\n",
              "      <td>4.152063</td>\n",
              "      <td>1.216834</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>ENSG00000146072</td>\n",
              "      <td>ENSG00000146072</td>\n",
              "      <td>6</td>\n",
              "      <td>47277641</td>\n",
              "      <td>2.567374</td>\n",
              "      <td>2.030769</td>\n",
              "      <td>0.927560</td>\n",
              "      <td>1.351749</td>\n",
              "      <td>2.316378</td>\n",
              "      <td>0.931071</td>\n",
              "      <td>...</td>\n",
              "      <td>0.761671</td>\n",
              "      <td>0.797139</td>\n",
              "      <td>0.328339</td>\n",
              "      <td>0.156504</td>\n",
              "      <td>1.118230</td>\n",
              "      <td>0.240833</td>\n",
              "      <td>1.022178</td>\n",
              "      <td>0.393550</td>\n",
              "      <td>0.007050</td>\n",
              "      <td>0.642450</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23712</th>\n",
              "      <td>ENSG00000087095</td>\n",
              "      <td>ENSG00000087095</td>\n",
              "      <td>17</td>\n",
              "      <td>26369182</td>\n",
              "      <td>5.490282</td>\n",
              "      <td>5.265422</td>\n",
              "      <td>3.457973</td>\n",
              "      <td>5.075550</td>\n",
              "      <td>5.571885</td>\n",
              "      <td>5.262668</td>\n",
              "      <td>...</td>\n",
              "      <td>6.678943</td>\n",
              "      <td>4.996638</td>\n",
              "      <td>5.125125</td>\n",
              "      <td>5.383470</td>\n",
              "      <td>5.206981</td>\n",
              "      <td>5.643264</td>\n",
              "      <td>5.290116</td>\n",
              "      <td>5.911672</td>\n",
              "      <td>5.249228</td>\n",
              "      <td>5.347071</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23714</th>\n",
              "      <td>ENSG00000162144</td>\n",
              "      <td>ENSG00000162144</td>\n",
              "      <td>11</td>\n",
              "      <td>61129771</td>\n",
              "      <td>33.050457</td>\n",
              "      <td>23.159734</td>\n",
              "      <td>16.050400</td>\n",
              "      <td>20.848751</td>\n",
              "      <td>32.245357</td>\n",
              "      <td>26.842346</td>\n",
              "      <td>...</td>\n",
              "      <td>22.962417</td>\n",
              "      <td>26.718492</td>\n",
              "      <td>17.096197</td>\n",
              "      <td>36.692807</td>\n",
              "      <td>32.645718</td>\n",
              "      <td>26.245967</td>\n",
              "      <td>25.040414</td>\n",
              "      <td>33.586088</td>\n",
              "      <td>27.436847</td>\n",
              "      <td>24.236840</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23715</th>\n",
              "      <td>ENSG00000129473</td>\n",
              "      <td>ENSG00000129473</td>\n",
              "      <td>14</td>\n",
              "      <td>23767999</td>\n",
              "      <td>5.960860</td>\n",
              "      <td>10.834208</td>\n",
              "      <td>4.606371</td>\n",
              "      <td>5.274462</td>\n",
              "      <td>6.453945</td>\n",
              "      <td>11.956208</td>\n",
              "      <td>...</td>\n",
              "      <td>7.947619</td>\n",
              "      <td>5.826466</td>\n",
              "      <td>6.691473</td>\n",
              "      <td>11.671064</td>\n",
              "      <td>9.606188</td>\n",
              "      <td>7.016099</td>\n",
              "      <td>6.305954</td>\n",
              "      <td>6.498428</td>\n",
              "      <td>4.919154</td>\n",
              "      <td>7.285334</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23718</th>\n",
              "      <td>ENSG00000114423</td>\n",
              "      <td>ENSG00000114423</td>\n",
              "      <td>3</td>\n",
              "      <td>105588396</td>\n",
              "      <td>14.054749</td>\n",
              "      <td>14.477899</td>\n",
              "      <td>11.584425</td>\n",
              "      <td>12.637956</td>\n",
              "      <td>12.015089</td>\n",
              "      <td>13.750655</td>\n",
              "      <td>...</td>\n",
              "      <td>11.723462</td>\n",
              "      <td>9.900372</td>\n",
              "      <td>10.473115</td>\n",
              "      <td>13.433413</td>\n",
              "      <td>15.832594</td>\n",
              "      <td>19.216176</td>\n",
              "      <td>10.213739</td>\n",
              "      <td>14.563192</td>\n",
              "      <td>15.637732</td>\n",
              "      <td>8.357117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23721</th>\n",
              "      <td>ENSG00000177494</td>\n",
              "      <td>ENSG00000177494</td>\n",
              "      <td>3</td>\n",
              "      <td>111314290</td>\n",
              "      <td>1.183398</td>\n",
              "      <td>0.892619</td>\n",
              "      <td>0.917497</td>\n",
              "      <td>3.504894</td>\n",
              "      <td>1.168247</td>\n",
              "      <td>3.110953</td>\n",
              "      <td>...</td>\n",
              "      <td>1.936493</td>\n",
              "      <td>1.177972</td>\n",
              "      <td>2.673129</td>\n",
              "      <td>13.892711</td>\n",
              "      <td>2.763749</td>\n",
              "      <td>3.032246</td>\n",
              "      <td>4.558199</td>\n",
              "      <td>2.807851</td>\n",
              "      <td>2.021329</td>\n",
              "      <td>3.181096</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>13536 rows × 466 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "              TargetID      Gene_Symbol Chr      Coord    HG00096    HG00097  \\\n",
              "1      ENSG00000183696  ENSG00000183696   7   48128225   8.183805   5.686911   \n",
              "2      ENSG00000139269  ENSG00000139269  12   57846106   1.199910   1.573572   \n",
              "3      ENSG00000169129  ENSG00000169129  10  116164515   0.831940   0.069778   \n",
              "5      ENSG00000136237  ENSG00000136237   7   22396763   3.788503   2.050963   \n",
              "13     ENSG00000146072  ENSG00000146072   6   47277641   2.567374   2.030769   \n",
              "...                ...              ...  ..        ...        ...        ...   \n",
              "23712  ENSG00000087095  ENSG00000087095  17   26369182   5.490282   5.265422   \n",
              "23714  ENSG00000162144  ENSG00000162144  11   61129771  33.050457  23.159734   \n",
              "23715  ENSG00000129473  ENSG00000129473  14   23767999   5.960860  10.834208   \n",
              "23718  ENSG00000114423  ENSG00000114423   3  105588396  14.054749  14.477899   \n",
              "23721  ENSG00000177494  ENSG00000177494   3  111314290   1.183398   0.892619   \n",
              "\n",
              "         HG00099    HG00100    HG00101    HG00102  ...    NA20810    NA20811  \\\n",
              "1       2.434653   3.830894   6.612288   4.709646  ...  13.428205   6.094500   \n",
              "2       0.521616   1.447225   3.565791   1.982681  ...   3.225880   1.996067   \n",
              "3       0.931086   0.620941   1.660668   0.570481  ...   1.023381   1.127852   \n",
              "5       4.000313   3.271619   1.798216   1.516688  ...   2.909393   1.921176   \n",
              "13      0.927560   1.351749   2.316378   0.931071  ...   0.761671   0.797139   \n",
              "...          ...        ...        ...        ...  ...        ...        ...   \n",
              "23712   3.457973   5.075550   5.571885   5.262668  ...   6.678943   4.996638   \n",
              "23714  16.050400  20.848751  32.245357  26.842346  ...  22.962417  26.718492   \n",
              "23715   4.606371   5.274462   6.453945  11.956208  ...   7.947619   5.826466   \n",
              "23718  11.584425  12.637956  12.015089  13.750655  ...  11.723462   9.900372   \n",
              "23721   0.917497   3.504894   1.168247   3.110953  ...   1.936493   1.177972   \n",
              "\n",
              "         NA20812    NA20813    NA20814    NA20815    NA20816    NA20819  \\\n",
              "1      12.536000   2.217262   3.573394   7.583364   4.052882   1.570378   \n",
              "2       2.854923   2.267343   1.331201   2.187895   1.004250   3.003316   \n",
              "3       0.774409   1.495854   0.895342   1.513521   0.826377   1.021201   \n",
              "5       5.083873   2.866573   1.297788   2.888316   2.145022   3.557598   \n",
              "13      0.328339   0.156504   1.118230   0.240833   1.022178   0.393550   \n",
              "...          ...        ...        ...        ...        ...        ...   \n",
              "23712   5.125125   5.383470   5.206981   5.643264   5.290116   5.911672   \n",
              "23714  17.096197  36.692807  32.645718  26.245967  25.040414  33.586088   \n",
              "23715   6.691473  11.671064   9.606188   7.016099   6.305954   6.498428   \n",
              "23718  10.473115  13.433413  15.832594  19.216176  10.213739  14.563192   \n",
              "23721   2.673129  13.892711   2.763749   3.032246   4.558199   2.807851   \n",
              "\n",
              "         NA20826    NA20828  \n",
              "1       4.900372   6.737308  \n",
              "2       1.984362   1.684954  \n",
              "3       0.952502   0.740565  \n",
              "5       4.152063   1.216834  \n",
              "13      0.007050   0.642450  \n",
              "...          ...        ...  \n",
              "23712   5.249228   5.347071  \n",
              "23714  27.436847  24.236840  \n",
              "23715   4.919154   7.285334  \n",
              "23718  15.637732   8.357117  \n",
              "23721   2.021329   3.181096  \n",
              "\n",
              "[13536 rows x 466 columns]"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "protein_coding_expression_data"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7bd6acd7",
      "metadata": {
        "id": "7bd6acd7"
      },
      "source": [
        "## Subset SNPs within +-500 Kb windows, and then conduct eQTL analysis for each of the 22 chromosomes"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "abe04250",
      "metadata": {
        "id": "abe04250"
      },
      "source": [
        "#### Create a dataframe that will incorporate all results of 22 chromosomes"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "a65c8824",
      "metadata": {
        "id": "a65c8824"
      },
      "outputs": [],
      "source": [
        "# Initialize an empty DataFrame to store results\n",
        "all_results = pd.DataFrame()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4609c5a8",
      "metadata": {
        "id": "4609c5a8"
      },
      "source": [
        "#### Chromosome 1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3cd1dd8d",
      "metadata": {
        "id": "3cd1dd8d",
        "outputId": "1f810685-f30d-425c-de9f-cd0eb0e1082d"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 10.94it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '1']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 1]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.1')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "adb1f289",
      "metadata": {
        "id": "adb1f289",
        "outputId": "6434a519-b610-410d-87f9-0a1bb16e9ce6"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ID</th>\n",
              "      <th>CHR</th>\n",
              "      <th>START</th>\n",
              "      <th>STOP</th>\n",
              "      <th>SYM</th>\n",
              "      <th>TYPE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>DDX11L1</td>\n",
              "      <td>1</td>\n",
              "      <td>11868</td>\n",
              "      <td>13052</td>\n",
              "      <td>ENSG00000223972</td>\n",
              "      <td>transcribed_unprocessed_pseudogene</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>OR4F5</td>\n",
              "      <td>1</td>\n",
              "      <td>69090</td>\n",
              "      <td>70008</td>\n",
              "      <td>ENSG00000186092</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FAM87B</td>\n",
              "      <td>1</td>\n",
              "      <td>817370</td>\n",
              "      <td>819834</td>\n",
              "      <td>ENSG00000177757</td>\n",
              "      <td>lincRNA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>LINC00115</td>\n",
              "      <td>1</td>\n",
              "      <td>826205</td>\n",
              "      <td>827522</td>\n",
              "      <td>ENSG00000225880</td>\n",
              "      <td>lincRNA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>LINC01128</td>\n",
              "      <td>1</td>\n",
              "      <td>827607</td>\n",
              "      <td>859446</td>\n",
              "      <td>ENSG00000228794</td>\n",
              "      <td>processed_transcript</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2139</th>\n",
              "      <td>SH3BP5L</td>\n",
              "      <td>1</td>\n",
              "      <td>248810448</td>\n",
              "      <td>248826633</td>\n",
              "      <td>ENSG00000175137</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2140</th>\n",
              "      <td>MIR3124</td>\n",
              "      <td>1</td>\n",
              "      <td>248826376</td>\n",
              "      <td>248826443</td>\n",
              "      <td>ENSG00000264500</td>\n",
              "      <td>miRNA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2141</th>\n",
              "      <td>ZNF672</td>\n",
              "      <td>1</td>\n",
              "      <td>248838209</td>\n",
              "      <td>248849517</td>\n",
              "      <td>ENSG00000171161</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2142</th>\n",
              "      <td>ZNF692</td>\n",
              "      <td>1</td>\n",
              "      <td>248850005</td>\n",
              "      <td>248859085</td>\n",
              "      <td>ENSG00000171163</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2143</th>\n",
              "      <td>PGBD2</td>\n",
              "      <td>1</td>\n",
              "      <td>248906195</td>\n",
              "      <td>248919946</td>\n",
              "      <td>ENSG00000185220</td>\n",
              "      <td>protein_coding</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>2144 rows × 6 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "             ID  CHR      START       STOP              SYM  \\\n",
              "0       DDX11L1    1      11868      13052  ENSG00000223972   \n",
              "1         OR4F5    1      69090      70008  ENSG00000186092   \n",
              "2        FAM87B    1     817370     819834  ENSG00000177757   \n",
              "3     LINC00115    1     826205     827522  ENSG00000225880   \n",
              "4     LINC01128    1     827607     859446  ENSG00000228794   \n",
              "...         ...  ...        ...        ...              ...   \n",
              "2139    SH3BP5L    1  248810448  248826633  ENSG00000175137   \n",
              "2140    MIR3124    1  248826376  248826443  ENSG00000264500   \n",
              "2141     ZNF672    1  248838209  248849517  ENSG00000171161   \n",
              "2142     ZNF692    1  248850005  248859085  ENSG00000171163   \n",
              "2143      PGBD2    1  248906195  248919946  ENSG00000185220   \n",
              "\n",
              "                                    TYPE  \n",
              "0     transcribed_unprocessed_pseudogene  \n",
              "1                         protein_coding  \n",
              "2                                lincRNA  \n",
              "3                                lincRNA  \n",
              "4                   processed_transcript  \n",
              "...                                  ...  \n",
              "2139                      protein_coding  \n",
              "2140                               miRNA  \n",
              "2141                      protein_coding  \n",
              "2142                      protein_coding  \n",
              "2143                      protein_coding  \n",
              "\n",
              "[2144 rows x 6 columns]"
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "gene_annot"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "21bad4af",
      "metadata": {
        "id": "21bad4af",
        "outputId": "22cf52c9-19d8-43cc-ee3d-d4c6b17c9ffa"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>TargetID</th>\n",
              "      <th>Gene_Symbol</th>\n",
              "      <th>Chr</th>\n",
              "      <th>Coord</th>\n",
              "      <th>HG00096</th>\n",
              "      <th>HG00097</th>\n",
              "      <th>HG00099</th>\n",
              "      <th>HG00100</th>\n",
              "      <th>HG00101</th>\n",
              "      <th>HG00102</th>\n",
              "      <th>...</th>\n",
              "      <th>NA20810</th>\n",
              "      <th>NA20811</th>\n",
              "      <th>NA20812</th>\n",
              "      <th>NA20813</th>\n",
              "      <th>NA20814</th>\n",
              "      <th>NA20815</th>\n",
              "      <th>NA20816</th>\n",
              "      <th>NA20819</th>\n",
              "      <th>NA20826</th>\n",
              "      <th>NA20828</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>ENSG00000183814</td>\n",
              "      <td>ENSG00000183814</td>\n",
              "      <td>1</td>\n",
              "      <td>226497570</td>\n",
              "      <td>9.324334</td>\n",
              "      <td>2.476283</td>\n",
              "      <td>2.109651</td>\n",
              "      <td>3.237022</td>\n",
              "      <td>4.656543</td>\n",
              "      <td>8.283848</td>\n",
              "      <td>...</td>\n",
              "      <td>7.454827</td>\n",
              "      <td>3.934654</td>\n",
              "      <td>8.519462</td>\n",
              "      <td>4.346162</td>\n",
              "      <td>4.106076</td>\n",
              "      <td>2.151307</td>\n",
              "      <td>2.534221</td>\n",
              "      <td>3.724759</td>\n",
              "      <td>3.224399</td>\n",
              "      <td>2.038018</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>ENSG00000163207</td>\n",
              "      <td>ENSG00000163207</td>\n",
              "      <td>1</td>\n",
              "      <td>152881021</td>\n",
              "      <td>2.319013</td>\n",
              "      <td>1.164032</td>\n",
              "      <td>3.214980</td>\n",
              "      <td>3.350841</td>\n",
              "      <td>4.690631</td>\n",
              "      <td>4.359018</td>\n",
              "      <td>...</td>\n",
              "      <td>3.549053</td>\n",
              "      <td>3.823884</td>\n",
              "      <td>2.299218</td>\n",
              "      <td>6.055119</td>\n",
              "      <td>6.852517</td>\n",
              "      <td>4.992477</td>\n",
              "      <td>3.607868</td>\n",
              "      <td>6.082010</td>\n",
              "      <td>1.770644</td>\n",
              "      <td>5.896630</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>64</th>\n",
              "      <td>ENSG00000137962</td>\n",
              "      <td>ENSG00000137962</td>\n",
              "      <td>1</td>\n",
              "      <td>94740624</td>\n",
              "      <td>0.129352</td>\n",
              "      <td>0.028266</td>\n",
              "      <td>0.035488</td>\n",
              "      <td>0.070250</td>\n",
              "      <td>0.074527</td>\n",
              "      <td>0.216915</td>\n",
              "      <td>...</td>\n",
              "      <td>0.110438</td>\n",
              "      <td>0.068898</td>\n",
              "      <td>0.049497</td>\n",
              "      <td>0.091802</td>\n",
              "      <td>0.066403</td>\n",
              "      <td>0.094006</td>\n",
              "      <td>0.031831</td>\n",
              "      <td>0.071078</td>\n",
              "      <td>0.106578</td>\n",
              "      <td>0.063013</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>72</th>\n",
              "      <td>ENSG00000117480</td>\n",
              "      <td>ENSG00000117480</td>\n",
              "      <td>1</td>\n",
              "      <td>46859937</td>\n",
              "      <td>0.354199</td>\n",
              "      <td>0.376160</td>\n",
              "      <td>0.325608</td>\n",
              "      <td>0.258863</td>\n",
              "      <td>0.472131</td>\n",
              "      <td>0.174365</td>\n",
              "      <td>...</td>\n",
              "      <td>0.649918</td>\n",
              "      <td>0.436623</td>\n",
              "      <td>0.768179</td>\n",
              "      <td>0.382980</td>\n",
              "      <td>0.568632</td>\n",
              "      <td>1.351411</td>\n",
              "      <td>0.766969</td>\n",
              "      <td>0.376778</td>\n",
              "      <td>0.374563</td>\n",
              "      <td>0.450959</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74</th>\n",
              "      <td>ENSG00000253368</td>\n",
              "      <td>ENSG00000253368</td>\n",
              "      <td>1</td>\n",
              "      <td>27320198</td>\n",
              "      <td>0.030788</td>\n",
              "      <td>0.133775</td>\n",
              "      <td>0.024396</td>\n",
              "      <td>0.075733</td>\n",
              "      <td>0.163359</td>\n",
              "      <td>0.106624</td>\n",
              "      <td>...</td>\n",
              "      <td>0.067799</td>\n",
              "      <td>0.059755</td>\n",
              "      <td>0.077219</td>\n",
              "      <td>0.057131</td>\n",
              "      <td>0.167180</td>\n",
              "      <td>0.046236</td>\n",
              "      <td>0.045522</td>\n",
              "      <td>0.053148</td>\n",
              "      <td>0.118932</td>\n",
              "      <td>0.205883</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23600</th>\n",
              "      <td>ENSG00000142794</td>\n",
              "      <td>ENSG00000142794</td>\n",
              "      <td>1</td>\n",
              "      <td>21766621</td>\n",
              "      <td>1.046401</td>\n",
              "      <td>0.906578</td>\n",
              "      <td>0.730067</td>\n",
              "      <td>1.638836</td>\n",
              "      <td>0.956008</td>\n",
              "      <td>2.388566</td>\n",
              "      <td>...</td>\n",
              "      <td>1.540532</td>\n",
              "      <td>0.226895</td>\n",
              "      <td>0.823943</td>\n",
              "      <td>0.389357</td>\n",
              "      <td>0.752907</td>\n",
              "      <td>0.305052</td>\n",
              "      <td>1.982624</td>\n",
              "      <td>0.210186</td>\n",
              "      <td>0.985346</td>\n",
              "      <td>0.396277</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23616</th>\n",
              "      <td>ENSG00000117009</td>\n",
              "      <td>ENSG00000117009</td>\n",
              "      <td>1</td>\n",
              "      <td>241695434</td>\n",
              "      <td>5.545435</td>\n",
              "      <td>4.914430</td>\n",
              "      <td>5.279781</td>\n",
              "      <td>4.936590</td>\n",
              "      <td>4.438147</td>\n",
              "      <td>4.727565</td>\n",
              "      <td>...</td>\n",
              "      <td>3.800012</td>\n",
              "      <td>4.438759</td>\n",
              "      <td>2.400835</td>\n",
              "      <td>4.907077</td>\n",
              "      <td>7.677483</td>\n",
              "      <td>5.968332</td>\n",
              "      <td>5.390299</td>\n",
              "      <td>5.578893</td>\n",
              "      <td>3.874899</td>\n",
              "      <td>3.992117</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23622</th>\n",
              "      <td>ENSG00000143514</td>\n",
              "      <td>ENSG00000143514</td>\n",
              "      <td>1</td>\n",
              "      <td>224033674</td>\n",
              "      <td>3.273854</td>\n",
              "      <td>2.317393</td>\n",
              "      <td>11.424262</td>\n",
              "      <td>4.673044</td>\n",
              "      <td>4.275406</td>\n",
              "      <td>2.841137</td>\n",
              "      <td>...</td>\n",
              "      <td>3.463870</td>\n",
              "      <td>4.075491</td>\n",
              "      <td>5.449698</td>\n",
              "      <td>3.058197</td>\n",
              "      <td>6.793001</td>\n",
              "      <td>6.495214</td>\n",
              "      <td>2.204188</td>\n",
              "      <td>3.800952</td>\n",
              "      <td>4.089346</td>\n",
              "      <td>6.165326</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23680</th>\n",
              "      <td>ENSG00000085491</td>\n",
              "      <td>ENSG00000085491</td>\n",
              "      <td>1</td>\n",
              "      <td>108743471</td>\n",
              "      <td>3.747062</td>\n",
              "      <td>1.568091</td>\n",
              "      <td>1.163357</td>\n",
              "      <td>0.747116</td>\n",
              "      <td>1.580397</td>\n",
              "      <td>0.988155</td>\n",
              "      <td>...</td>\n",
              "      <td>1.189579</td>\n",
              "      <td>1.770024</td>\n",
              "      <td>0.898182</td>\n",
              "      <td>2.338589</td>\n",
              "      <td>0.728569</td>\n",
              "      <td>3.337545</td>\n",
              "      <td>3.240567</td>\n",
              "      <td>0.651865</td>\n",
              "      <td>0.593732</td>\n",
              "      <td>0.015042</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23702</th>\n",
              "      <td>ENSG00000171502</td>\n",
              "      <td>ENSG00000171502</td>\n",
              "      <td>1</td>\n",
              "      <td>86622626</td>\n",
              "      <td>0.560208</td>\n",
              "      <td>1.345523</td>\n",
              "      <td>0.487027</td>\n",
              "      <td>0.932744</td>\n",
              "      <td>0.100971</td>\n",
              "      <td>2.034533</td>\n",
              "      <td>...</td>\n",
              "      <td>0.590634</td>\n",
              "      <td>1.084875</td>\n",
              "      <td>0.744239</td>\n",
              "      <td>1.218765</td>\n",
              "      <td>0.296234</td>\n",
              "      <td>1.036922</td>\n",
              "      <td>0.991679</td>\n",
              "      <td>2.001817</td>\n",
              "      <td>0.678879</td>\n",
              "      <td>0.579966</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1467 rows × 466 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "              TargetID      Gene_Symbol Chr      Coord   HG00096   HG00097  \\\n",
              "14     ENSG00000183814  ENSG00000183814   1  226497570  9.324334  2.476283   \n",
              "17     ENSG00000163207  ENSG00000163207   1  152881021  2.319013  1.164032   \n",
              "64     ENSG00000137962  ENSG00000137962   1   94740624  0.129352  0.028266   \n",
              "72     ENSG00000117480  ENSG00000117480   1   46859937  0.354199  0.376160   \n",
              "74     ENSG00000253368  ENSG00000253368   1   27320198  0.030788  0.133775   \n",
              "...                ...              ...  ..        ...       ...       ...   \n",
              "23600  ENSG00000142794  ENSG00000142794   1   21766621  1.046401  0.906578   \n",
              "23616  ENSG00000117009  ENSG00000117009   1  241695434  5.545435  4.914430   \n",
              "23622  ENSG00000143514  ENSG00000143514   1  224033674  3.273854  2.317393   \n",
              "23680  ENSG00000085491  ENSG00000085491   1  108743471  3.747062  1.568091   \n",
              "23702  ENSG00000171502  ENSG00000171502   1   86622626  0.560208  1.345523   \n",
              "\n",
              "         HG00099   HG00100   HG00101   HG00102  ...   NA20810   NA20811  \\\n",
              "14      2.109651  3.237022  4.656543  8.283848  ...  7.454827  3.934654   \n",
              "17      3.214980  3.350841  4.690631  4.359018  ...  3.549053  3.823884   \n",
              "64      0.035488  0.070250  0.074527  0.216915  ...  0.110438  0.068898   \n",
              "72      0.325608  0.258863  0.472131  0.174365  ...  0.649918  0.436623   \n",
              "74      0.024396  0.075733  0.163359  0.106624  ...  0.067799  0.059755   \n",
              "...          ...       ...       ...       ...  ...       ...       ...   \n",
              "23600   0.730067  1.638836  0.956008  2.388566  ...  1.540532  0.226895   \n",
              "23616   5.279781  4.936590  4.438147  4.727565  ...  3.800012  4.438759   \n",
              "23622  11.424262  4.673044  4.275406  2.841137  ...  3.463870  4.075491   \n",
              "23680   1.163357  0.747116  1.580397  0.988155  ...  1.189579  1.770024   \n",
              "23702   0.487027  0.932744  0.100971  2.034533  ...  0.590634  1.084875   \n",
              "\n",
              "        NA20812   NA20813   NA20814   NA20815   NA20816   NA20819   NA20826  \\\n",
              "14     8.519462  4.346162  4.106076  2.151307  2.534221  3.724759  3.224399   \n",
              "17     2.299218  6.055119  6.852517  4.992477  3.607868  6.082010  1.770644   \n",
              "64     0.049497  0.091802  0.066403  0.094006  0.031831  0.071078  0.106578   \n",
              "72     0.768179  0.382980  0.568632  1.351411  0.766969  0.376778  0.374563   \n",
              "74     0.077219  0.057131  0.167180  0.046236  0.045522  0.053148  0.118932   \n",
              "...         ...       ...       ...       ...       ...       ...       ...   \n",
              "23600  0.823943  0.389357  0.752907  0.305052  1.982624  0.210186  0.985346   \n",
              "23616  2.400835  4.907077  7.677483  5.968332  5.390299  5.578893  3.874899   \n",
              "23622  5.449698  3.058197  6.793001  6.495214  2.204188  3.800952  4.089346   \n",
              "23680  0.898182  2.338589  0.728569  3.337545  3.240567  0.651865  0.593732   \n",
              "23702  0.744239  1.218765  0.296234  1.036922  0.991679  2.001817  0.678879   \n",
              "\n",
              "        NA20828  \n",
              "14     2.038018  \n",
              "17     5.896630  \n",
              "64     0.063013  \n",
              "72     0.450959  \n",
              "74     0.205883  \n",
              "...         ...  \n",
              "23600  0.396277  \n",
              "23616  3.992117  \n",
              "23622  6.165326  \n",
              "23680  0.015042  \n",
              "23702  0.579966  \n",
              "\n",
              "[1467 rows x 466 columns]"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "expression_data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "8ea0f441",
      "metadata": {
        "id": "8ea0f441"
      },
      "outputs": [],
      "source": [
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7fd147a9",
      "metadata": {
        "id": "7fd147a9"
      },
      "source": [
        "#### Chromosome 2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "582f73c3",
      "metadata": {
        "id": "582f73c3",
        "outputId": "cc8532fd-2bfd-4c0a-cdd3-74a5f5f18ff3"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  9.13it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '2']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 2]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.2')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5cc8d6c1",
      "metadata": {
        "id": "5cc8d6c1"
      },
      "source": [
        "#### Chromosome 3"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "28ff2426",
      "metadata": {
        "id": "28ff2426",
        "outputId": "251dc264-8c6d-4e6d-f1bb-e18f370374e9"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 10.32it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '3']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 3]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.3')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "87a4abbf",
      "metadata": {
        "id": "87a4abbf"
      },
      "source": [
        "#### Chromosome 4"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "cfba72e1",
      "metadata": {
        "id": "cfba72e1",
        "outputId": "4c8430d2-6ba5-41f3-a8c4-f392d25c5178"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  8.99it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '4']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 4]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.4')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "7e88c472",
      "metadata": {
        "id": "7e88c472"
      },
      "source": [
        "#### Chromosome 5"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "a46171c2",
      "metadata": {
        "id": "a46171c2",
        "outputId": "4090c637-2e6d-4047-94ba-0ca36a16bfe0"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 14.60it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '5']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 5]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.5')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "797672a4",
      "metadata": {
        "id": "797672a4"
      },
      "source": [
        "#### Chromosome 6"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1e353082",
      "metadata": {
        "id": "1e353082",
        "outputId": "89ad4b43-3a2f-4d84-aa8f-cfead0dcacea"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  9.89it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '6']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 6]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.6')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "d90f2f9a",
      "metadata": {
        "id": "d90f2f9a"
      },
      "source": [
        "#### Chromosome 7"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "1ddfa4ac",
      "metadata": {
        "id": "1ddfa4ac",
        "outputId": "d56fbeda-d303-40de-ff89-85d1db841aa3"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 13.69it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '7']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 7]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.7')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "845ee03f",
      "metadata": {
        "id": "845ee03f"
      },
      "source": [
        "#### Chromosome 8"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3c208bf8",
      "metadata": {
        "id": "3c208bf8",
        "outputId": "3128d3e4-7401-460d-c112-d111cac8bfc6"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 13.37it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '8']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 8]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.8')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "db94282d",
      "metadata": {
        "id": "db94282d"
      },
      "source": [
        "#### Chromosome 9"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "130f46c7",
      "metadata": {
        "id": "130f46c7",
        "outputId": "913c985f-2e0f-4ddb-e64e-5499be91fb5b"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 22.89it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '9']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 9]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.9')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "fa20cb72",
      "metadata": {
        "id": "fa20cb72"
      },
      "source": [
        "#### Chromosome 10"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5fa2bece",
      "metadata": {
        "id": "5fa2bece",
        "outputId": "2ba75eb4-4e71-4ebc-a910-fc3986cc250c"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 16.88it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '10']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 10]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.10')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "621de57d",
      "metadata": {
        "id": "621de57d"
      },
      "source": [
        "#### Chromosome 11"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "88ea37e3",
      "metadata": {
        "id": "88ea37e3",
        "outputId": "1164719c-3c48-4f6f-a9b0-3a20488f752b"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 11.12it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '11']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 11]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.11')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "57f5ddb6",
      "metadata": {
        "id": "57f5ddb6"
      },
      "source": [
        "#### Chromosome 12"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "c63f0cca",
      "metadata": {
        "id": "c63f0cca",
        "outputId": "1948dda7-5fdd-461b-ede4-2cce8f55f9f1"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 15.73it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '12']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 12]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.12')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "5a89b852",
      "metadata": {
        "id": "5a89b852"
      },
      "source": [
        "#### Chromosome 13"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ae29c4a8",
      "metadata": {
        "id": "ae29c4a8",
        "outputId": "56929374-abe6-4907-fe72-30aa554997b9"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 17.70it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '13']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 13]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.13')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0b5541b7",
      "metadata": {
        "id": "0b5541b7"
      },
      "source": [
        "#### Chromosome 14"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f5795bb7",
      "metadata": {
        "id": "f5795bb7",
        "outputId": "a65b9c40-6a34-4040-db4f-4a00ed7a504f"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 17.52it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '14']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 14]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.14')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a7b52b06",
      "metadata": {
        "id": "a7b52b06"
      },
      "source": [
        "#### Chromosome 15"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4e2cbcc0",
      "metadata": {
        "id": "4e2cbcc0",
        "outputId": "2c044441-943c-4508-f500-89b46860d1a1"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 18.15it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '15']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 15]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.15')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "b08eacfd",
      "metadata": {
        "id": "b08eacfd"
      },
      "source": [
        "#### Chromosome 16"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "4401d70d",
      "metadata": {
        "id": "4401d70d",
        "outputId": "606f5bc5-11ff-48b2-c1ac-dafe92def0f1"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 15.88it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '16']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 16]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.16')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4851b7fc",
      "metadata": {
        "id": "4851b7fc"
      },
      "source": [
        "#### Chromosome 17"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3e53bf09",
      "metadata": {
        "id": "3e53bf09",
        "outputId": "30d43086-d175-4db4-cd7a-960ee7b17588"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 21.79it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '17']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 17]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.17')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ed77e611",
      "metadata": {
        "id": "ed77e611"
      },
      "source": [
        "#### Chromosome 18"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e4e8456b",
      "metadata": {
        "id": "e4e8456b",
        "outputId": "b9dde63b-784b-46f6-d0c9-e6b26289445a"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 19.46it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '18']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 18]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.18')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "28a42085",
      "metadata": {
        "id": "28a42085"
      },
      "source": [
        "#### Chromosome 19"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "fcd6aa6a",
      "metadata": {
        "id": "fcd6aa6a",
        "outputId": "ee493c01-ef55-4296-d976-6395dd43bb7e"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 24.30it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '19']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 19]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.19')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e2be185e",
      "metadata": {
        "id": "e2be185e"
      },
      "source": [
        "#### Chromosome 20"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "43b07b17",
      "metadata": {
        "id": "43b07b17",
        "outputId": "da817649-8500-46d0-f455-d9bb0e460d15"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 28.58it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '20']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 20]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.20')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f2cab4e6",
      "metadata": {
        "id": "f2cab4e6"
      },
      "source": [
        "#### Chromosome 21"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "23cb7d39",
      "metadata": {
        "id": "23cb7d39",
        "outputId": "e64fadf1-157f-4fe9-c5bc-e1eea0c563e6"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 25.47it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '21']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 21]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.21')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ad140aaf",
      "metadata": {
        "id": "ad140aaf"
      },
      "source": [
        "#### Chromosome 22"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "ecf917eb",
      "metadata": {
        "id": "ecf917eb",
        "outputId": "f01d8bb8-b156-48ad-870a-7457438116e4"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 23.06it/s]\n"
          ]
        }
      ],
      "source": [
        "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '22']\n",
        "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 22]\n",
        "\n",
        "# Read plink files\n",
        "(bim, fam, bed) = pp.read_plink('1000G.EUR.22')\n",
        "\n",
        "#Find common individuals\n",
        "fam_ids = fam['fid'].tolist()\n",
        "expression_ids = expression_data.columns.tolist()\n",
        "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
        "intersection_ids_list = list(intersection_ids)\n",
        "\n",
        "# Filter the expression data and genotype data to include only common individuals\n",
        "expression_data_filtered = expression_data[intersection_ids_list]\n",
        "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
        "\n",
        "#Cis-eQTL Analysis\n",
        "results = []\n",
        "window_size = 500000  # 500 Kb window size\n",
        "for gene_row in gene_annot.itertuples():\n",
        "    gene_id = gene_row.ID\n",
        "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
        "    gene_start = gene_row.START\n",
        "    gene_end = gene_row.STOP\n",
        "\n",
        "    # Filter SNPs within the window of the gene\n",
        "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
        "\n",
        "    if cis_snps.empty:\n",
        "        continue\n",
        "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
        "\n",
        "    if gene_expression_df.empty:\n",
        "        continue\n",
        "\n",
        "    gene_expression = gene_expression_df.iloc[0]\n",
        "\n",
        "    for snp_row in cis_snps.itertuples():\n",
        "        snp_id = snp_row.snp\n",
        "        snp = bed_filtered[snp_row.Index, :]\n",
        "\n",
        "        if len(gene_expression) != len(snp):\n",
        "            # Handle mismatch in length between gene expression and SNP data\n",
        "            continue\n",
        "\n",
        "        # Add a constant term to the independent variables to represent the intercept\n",
        "        snp_with_intercept = sm.add_constant(snp)\n",
        "\n",
        "        # Perform linear regression\n",
        "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
        "        result = model.fit()\n",
        "\n",
        "        # Check if there are enough parameters in the results\n",
        "        if len(result.params) < 2:\n",
        "            continue\n",
        "\n",
        "        pvalue = result.pvalues[1]\n",
        "        beta = result.params[1]\n",
        "        std_err = result.bse[1]\n",
        "\n",
        "        results.append({\n",
        "            'gene_id': gene_id,\n",
        "            'gene_symbol': gene_sym,\n",
        "            'snp_id': snp_id,\n",
        "            'pvalue': pvalue,\n",
        "            'OR': beta,\n",
        "            'SE': std_err,\n",
        "            'snp_position': snp_row.pos\n",
        "        })\n",
        "\n",
        "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f44b3cce",
      "metadata": {
        "id": "f44b3cce"
      },
      "source": [
        "### Total results of all 22 chromosomes"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "39ee0d65",
      "metadata": {
        "id": "39ee0d65",
        "outputId": "8c743753-cbb2-49b9-cef5-3cc2486dd0ee"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_symbol</th>\n",
              "      <th>snp_id</th>\n",
              "      <th>pvalue</th>\n",
              "      <th>OR</th>\n",
              "      <th>SE</th>\n",
              "      <th>snp_position</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3094315</td>\n",
              "      <td>0.298244</td>\n",
              "      <td>0.094348</td>\n",
              "      <td>0.090563</td>\n",
              "      <td>752566</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3131972</td>\n",
              "      <td>0.276524</td>\n",
              "      <td>0.098526</td>\n",
              "      <td>0.090400</td>\n",
              "      <td>752721</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3131969</td>\n",
              "      <td>0.538234</td>\n",
              "      <td>0.058996</td>\n",
              "      <td>0.095756</td>\n",
              "      <td>754182</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs1048488</td>\n",
              "      <td>0.317524</td>\n",
              "      <td>0.089695</td>\n",
              "      <td>0.089604</td>\n",
              "      <td>760912</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3115850</td>\n",
              "      <td>0.302921</td>\n",
              "      <td>0.092290</td>\n",
              "      <td>0.089450</td>\n",
              "      <td>761147</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183028</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs13056621</td>\n",
              "      <td>0.850607</td>\n",
              "      <td>-0.051107</td>\n",
              "      <td>0.271142</td>\n",
              "      <td>51181759</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183029</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs3865766</td>\n",
              "      <td>0.322009</td>\n",
              "      <td>-0.178194</td>\n",
              "      <td>0.179672</td>\n",
              "      <td>51186228</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183030</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs3888396</td>\n",
              "      <td>0.511601</td>\n",
              "      <td>0.164925</td>\n",
              "      <td>0.251014</td>\n",
              "      <td>51211392</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183031</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs2238837</td>\n",
              "      <td>0.338360</td>\n",
              "      <td>-0.184742</td>\n",
              "      <td>0.192690</td>\n",
              "      <td>51212875</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183032</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs28729663</td>\n",
              "      <td>0.726480</td>\n",
              "      <td>0.089485</td>\n",
              "      <td>0.255598</td>\n",
              "      <td>51219006</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6183033 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        gene_id      gene_symbol      snp_id    pvalue        OR        SE  \\\n",
              "0        SAMD11  ENSG00000187634   rs3094315  0.298244  0.094348  0.090563   \n",
              "1        SAMD11  ENSG00000187634   rs3131972  0.276524  0.098526  0.090400   \n",
              "2        SAMD11  ENSG00000187634   rs3131969  0.538234  0.058996  0.095756   \n",
              "3        SAMD11  ENSG00000187634   rs1048488  0.317524  0.089695  0.089604   \n",
              "4        SAMD11  ENSG00000187634   rs3115850  0.302921  0.092290  0.089450   \n",
              "...         ...              ...         ...       ...       ...       ...   \n",
              "6183028  RABL2B  ENSG00000079974  rs13056621  0.850607 -0.051107  0.271142   \n",
              "6183029  RABL2B  ENSG00000079974   rs3865766  0.322009 -0.178194  0.179672   \n",
              "6183030  RABL2B  ENSG00000079974   rs3888396  0.511601  0.164925  0.251014   \n",
              "6183031  RABL2B  ENSG00000079974   rs2238837  0.338360 -0.184742  0.192690   \n",
              "6183032  RABL2B  ENSG00000079974  rs28729663  0.726480  0.089485  0.255598   \n",
              "\n",
              "         snp_position  \n",
              "0              752566  \n",
              "1              752721  \n",
              "2              754182  \n",
              "3              760912  \n",
              "4              761147  \n",
              "...               ...  \n",
              "6183028      51181759  \n",
              "6183029      51186228  \n",
              "6183030      51211392  \n",
              "6183031      51212875  \n",
              "6183032      51219006  \n",
              "\n",
              "[6183033 rows x 7 columns]"
            ]
          },
          "execution_count": 47,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "all_results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "fa377be2",
      "metadata": {
        "id": "fa377be2"
      },
      "outputs": [],
      "source": [
        "max_rows = 1048576  # Excel's limit\n",
        "num_sections = len(all_results) // max_rows + 1\n",
        "\n",
        "# Create a Pandas Excel writer using XlsxWriter as the engine\n",
        "with pd.ExcelWriter('large_file.xlsx', engine='xlsxwriter') as writer:\n",
        "    for section in range(num_sections):\n",
        "        start_row = max_rows * section\n",
        "        end_row = start_row + max_rows\n",
        "        # Write each section of the DataFrame to a different sheet\n",
        "        all_results.iloc[start_row:end_row].to_excel(writer, sheet_name=f'Project1_eQTL_results{section + 1}', index=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "7417ba01",
      "metadata": {
        "id": "7417ba01"
      },
      "outputs": [],
      "source": [
        "file_pattern = '*.xlsx'  # Assumes files have .xlsx extension\n",
        "\n",
        "# Create a list of all Excel files in the directory\n",
        "excel_files = glob.glob(file_pattern)\n",
        "\n",
        "# List of sheet names to read\n",
        "sheet_names = ['Project1_eQTL_results1', 'Project1_eQTL_results2', 'Project1_eQTL_results3',\n",
        "               'Project1_eQTL_results4', 'Project1_eQTL_results5', 'Project1_eQTL_results6']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d170c4dc",
      "metadata": {
        "id": "d170c4dc",
        "outputId": "3106a21a-45d9-4d89-bf71-4da5c7e9e3bf"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['large_file.xlsx']"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "excel_files"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e44a4fb4",
      "metadata": {
        "id": "e44a4fb4"
      },
      "outputs": [],
      "source": [
        "# Read each sheet in each Excel file and store in a list\n",
        "dfs = []\n",
        "for file in excel_files:\n",
        "    for sheet in sheet_names:\n",
        "        df = pd.read_excel(file, sheet_name=sheet)\n",
        "        dfs.append(df)\n",
        "saved_results = pd.concat(dfs, ignore_index=True)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "af2bf277",
      "metadata": {
        "id": "af2bf277",
        "outputId": "0419a12f-de8f-415d-c655-635b28242539"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_symbol</th>\n",
              "      <th>snp_id</th>\n",
              "      <th>pvalue</th>\n",
              "      <th>OR</th>\n",
              "      <th>SE</th>\n",
              "      <th>snp_position</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3094315</td>\n",
              "      <td>0.298244</td>\n",
              "      <td>0.094348</td>\n",
              "      <td>0.090563</td>\n",
              "      <td>752566</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3131972</td>\n",
              "      <td>0.276524</td>\n",
              "      <td>0.098526</td>\n",
              "      <td>0.090400</td>\n",
              "      <td>752721</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3131969</td>\n",
              "      <td>0.538234</td>\n",
              "      <td>0.058996</td>\n",
              "      <td>0.095756</td>\n",
              "      <td>754182</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs1048488</td>\n",
              "      <td>0.317524</td>\n",
              "      <td>0.089695</td>\n",
              "      <td>0.089604</td>\n",
              "      <td>760912</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>rs3115850</td>\n",
              "      <td>0.302921</td>\n",
              "      <td>0.092290</td>\n",
              "      <td>0.089450</td>\n",
              "      <td>761147</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183023</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs13056621</td>\n",
              "      <td>0.850607</td>\n",
              "      <td>-0.051107</td>\n",
              "      <td>0.271142</td>\n",
              "      <td>51181759</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183024</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs3865766</td>\n",
              "      <td>0.322009</td>\n",
              "      <td>-0.178194</td>\n",
              "      <td>0.179672</td>\n",
              "      <td>51186228</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183025</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs3888396</td>\n",
              "      <td>0.511601</td>\n",
              "      <td>0.164925</td>\n",
              "      <td>0.251014</td>\n",
              "      <td>51211392</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183026</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs2238837</td>\n",
              "      <td>0.338360</td>\n",
              "      <td>-0.184742</td>\n",
              "      <td>0.192690</td>\n",
              "      <td>51212875</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6183027</th>\n",
              "      <td>RABL2B</td>\n",
              "      <td>ENSG00000079974</td>\n",
              "      <td>rs28729663</td>\n",
              "      <td>0.726480</td>\n",
              "      <td>0.089485</td>\n",
              "      <td>0.255598</td>\n",
              "      <td>51219006</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6183028 rows × 7 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "        gene_id      gene_symbol      snp_id    pvalue        OR        SE  \\\n",
              "0        SAMD11  ENSG00000187634   rs3094315  0.298244  0.094348  0.090563   \n",
              "1        SAMD11  ENSG00000187634   rs3131972  0.276524  0.098526  0.090400   \n",
              "2        SAMD11  ENSG00000187634   rs3131969  0.538234  0.058996  0.095756   \n",
              "3        SAMD11  ENSG00000187634   rs1048488  0.317524  0.089695  0.089604   \n",
              "4        SAMD11  ENSG00000187634   rs3115850  0.302921  0.092290  0.089450   \n",
              "...         ...              ...         ...       ...       ...       ...   \n",
              "6183023  RABL2B  ENSG00000079974  rs13056621  0.850607 -0.051107  0.271142   \n",
              "6183024  RABL2B  ENSG00000079974   rs3865766  0.322009 -0.178194  0.179672   \n",
              "6183025  RABL2B  ENSG00000079974   rs3888396  0.511601  0.164925  0.251014   \n",
              "6183026  RABL2B  ENSG00000079974   rs2238837  0.338360 -0.184742  0.192690   \n",
              "6183027  RABL2B  ENSG00000079974  rs28729663  0.726480  0.089485  0.255598   \n",
              "\n",
              "         snp_position  \n",
              "0              752566  \n",
              "1              752721  \n",
              "2              754182  \n",
              "3              760912  \n",
              "4              761147  \n",
              "...               ...  \n",
              "6183023      51181759  \n",
              "6183024      51186228  \n",
              "6183025      51211392  \n",
              "6183026      51212875  \n",
              "6183027      51219006  \n",
              "\n",
              "[6183028 rows x 7 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "saved_results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "316c1775",
      "metadata": {
        "id": "316c1775"
      },
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.8"
    },
    "colab": {
      "provenance": [],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
