{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9cfa2da9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas_plink as pp\n",
    "import statsmodels.api as sm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55388f50",
   "metadata": {},
   "source": [
    "#### Take a look at the gene_expression, gene_annotation files we have, we will subset gene expression data to find genes that are protein coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0e649747",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the data folder\n",
    "data_folder = '../data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bcb5475f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_file = data_folder + 'GD462.GeneQuantRPKM.50FN.samplename.resk10.txt'\n",
    "gene_expression_data = pd.read_csv(gene_expression_file, delimiter='\\t')\n",
    "\n",
    "#slice the string to make the gene symbols consistent with gene_annotation data\n",
    "gene_expression_data['TargetID'] = gene_expression_data['TargetID'].str.split('.').str[0]\n",
    "gene_expression_data['Gene_Symbol'] = gene_expression_data['Gene_Symbol'].str.split('.').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "726087b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TargetID</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Coord</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>...</th>\n",
       "      <th>NA20810</th>\n",
       "      <th>NA20811</th>\n",
       "      <th>NA20812</th>\n",
       "      <th>NA20813</th>\n",
       "      <th>NA20814</th>\n",
       "      <th>NA20815</th>\n",
       "      <th>NA20816</th>\n",
       "      <th>NA20819</th>\n",
       "      <th>NA20826</th>\n",
       "      <th>NA20828</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000152931</td>\n",
       "      <td>ENSG00000152931</td>\n",
       "      <td>5</td>\n",
       "      <td>59783540</td>\n",
       "      <td>0.101858</td>\n",
       "      <td>0.078110</td>\n",
       "      <td>0.048981</td>\n",
       "      <td>0.118597</td>\n",
       "      <td>0.004035</td>\n",
       "      <td>0.010925</td>\n",
       "      <td>...</td>\n",
       "      <td>0.088601</td>\n",
       "      <td>0.240010</td>\n",
       "      <td>0.137175</td>\n",
       "      <td>0.148494</td>\n",
       "      <td>0.038643</td>\n",
       "      <td>0.088509</td>\n",
       "      <td>0.029204</td>\n",
       "      <td>0.024423</td>\n",
       "      <td>0.044816</td>\n",
       "      <td>0.139186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000183696</td>\n",
       "      <td>ENSG00000183696</td>\n",
       "      <td>7</td>\n",
       "      <td>48128225</td>\n",
       "      <td>8.183805</td>\n",
       "      <td>5.686911</td>\n",
       "      <td>2.434653</td>\n",
       "      <td>3.830894</td>\n",
       "      <td>6.612288</td>\n",
       "      <td>4.709646</td>\n",
       "      <td>...</td>\n",
       "      <td>13.428205</td>\n",
       "      <td>6.094500</td>\n",
       "      <td>12.536000</td>\n",
       "      <td>2.217262</td>\n",
       "      <td>3.573394</td>\n",
       "      <td>7.583364</td>\n",
       "      <td>4.052882</td>\n",
       "      <td>1.570378</td>\n",
       "      <td>4.900372</td>\n",
       "      <td>6.737308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000139269</td>\n",
       "      <td>ENSG00000139269</td>\n",
       "      <td>12</td>\n",
       "      <td>57846106</td>\n",
       "      <td>1.199910</td>\n",
       "      <td>1.573572</td>\n",
       "      <td>0.521616</td>\n",
       "      <td>1.447225</td>\n",
       "      <td>3.565791</td>\n",
       "      <td>1.982681</td>\n",
       "      <td>...</td>\n",
       "      <td>3.225880</td>\n",
       "      <td>1.996067</td>\n",
       "      <td>2.854923</td>\n",
       "      <td>2.267343</td>\n",
       "      <td>1.331201</td>\n",
       "      <td>2.187895</td>\n",
       "      <td>1.004250</td>\n",
       "      <td>3.003316</td>\n",
       "      <td>1.984362</td>\n",
       "      <td>1.684954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000169129</td>\n",
       "      <td>ENSG00000169129</td>\n",
       "      <td>10</td>\n",
       "      <td>116164515</td>\n",
       "      <td>0.831940</td>\n",
       "      <td>0.069778</td>\n",
       "      <td>0.931086</td>\n",
       "      <td>0.620941</td>\n",
       "      <td>1.660668</td>\n",
       "      <td>0.570481</td>\n",
       "      <td>...</td>\n",
       "      <td>1.023381</td>\n",
       "      <td>1.127852</td>\n",
       "      <td>0.774409</td>\n",
       "      <td>1.495854</td>\n",
       "      <td>0.895342</td>\n",
       "      <td>1.513521</td>\n",
       "      <td>0.826377</td>\n",
       "      <td>1.021201</td>\n",
       "      <td>0.952502</td>\n",
       "      <td>0.740565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000134602</td>\n",
       "      <td>ENSG00000134602</td>\n",
       "      <td>X</td>\n",
       "      <td>131157293</td>\n",
       "      <td>27.646422</td>\n",
       "      <td>24.395572</td>\n",
       "      <td>16.445374</td>\n",
       "      <td>24.806650</td>\n",
       "      <td>25.113349</td>\n",
       "      <td>19.233988</td>\n",
       "      <td>...</td>\n",
       "      <td>25.079490</td>\n",
       "      <td>28.725528</td>\n",
       "      <td>24.450520</td>\n",
       "      <td>27.264069</td>\n",
       "      <td>26.912814</td>\n",
       "      <td>29.509210</td>\n",
       "      <td>26.462331</td>\n",
       "      <td>25.624009</td>\n",
       "      <td>25.707741</td>\n",
       "      <td>22.824957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23717</th>\n",
       "      <td>ENSG00000235472</td>\n",
       "      <td>ENSG00000235472</td>\n",
       "      <td>13</td>\n",
       "      <td>29172970</td>\n",
       "      <td>31.582832</td>\n",
       "      <td>34.071123</td>\n",
       "      <td>19.394365</td>\n",
       "      <td>37.523721</td>\n",
       "      <td>33.430473</td>\n",
       "      <td>37.844966</td>\n",
       "      <td>...</td>\n",
       "      <td>47.905182</td>\n",
       "      <td>33.224977</td>\n",
       "      <td>39.827675</td>\n",
       "      <td>27.096811</td>\n",
       "      <td>34.686342</td>\n",
       "      <td>37.386766</td>\n",
       "      <td>30.598840</td>\n",
       "      <td>33.516674</td>\n",
       "      <td>32.017940</td>\n",
       "      <td>38.341888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23718</th>\n",
       "      <td>ENSG00000114423</td>\n",
       "      <td>ENSG00000114423</td>\n",
       "      <td>3</td>\n",
       "      <td>105588396</td>\n",
       "      <td>14.054749</td>\n",
       "      <td>14.477899</td>\n",
       "      <td>11.584425</td>\n",
       "      <td>12.637956</td>\n",
       "      <td>12.015089</td>\n",
       "      <td>13.750655</td>\n",
       "      <td>...</td>\n",
       "      <td>11.723462</td>\n",
       "      <td>9.900372</td>\n",
       "      <td>10.473115</td>\n",
       "      <td>13.433413</td>\n",
       "      <td>15.832594</td>\n",
       "      <td>19.216176</td>\n",
       "      <td>10.213739</td>\n",
       "      <td>14.563192</td>\n",
       "      <td>15.637732</td>\n",
       "      <td>8.357117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23719</th>\n",
       "      <td>ENSG00000243312</td>\n",
       "      <td>ENSG00000243312</td>\n",
       "      <td>4</td>\n",
       "      <td>87791344</td>\n",
       "      <td>1.112114</td>\n",
       "      <td>0.831797</td>\n",
       "      <td>0.253228</td>\n",
       "      <td>0.271568</td>\n",
       "      <td>0.486086</td>\n",
       "      <td>1.362640</td>\n",
       "      <td>...</td>\n",
       "      <td>1.168991</td>\n",
       "      <td>0.645389</td>\n",
       "      <td>0.819469</td>\n",
       "      <td>0.515448</td>\n",
       "      <td>0.463054</td>\n",
       "      <td>1.580658</td>\n",
       "      <td>0.701396</td>\n",
       "      <td>0.771233</td>\n",
       "      <td>0.857330</td>\n",
       "      <td>0.703369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23720</th>\n",
       "      <td>ENSG00000257337</td>\n",
       "      <td>ENSG00000257337</td>\n",
       "      <td>12</td>\n",
       "      <td>53448222</td>\n",
       "      <td>3.826396</td>\n",
       "      <td>6.045798</td>\n",
       "      <td>2.593872</td>\n",
       "      <td>4.447169</td>\n",
       "      <td>5.294657</td>\n",
       "      <td>4.106823</td>\n",
       "      <td>...</td>\n",
       "      <td>6.524398</td>\n",
       "      <td>6.017795</td>\n",
       "      <td>3.460273</td>\n",
       "      <td>3.905032</td>\n",
       "      <td>5.023161</td>\n",
       "      <td>5.333027</td>\n",
       "      <td>5.694370</td>\n",
       "      <td>8.142939</td>\n",
       "      <td>5.622043</td>\n",
       "      <td>6.026476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23721</th>\n",
       "      <td>ENSG00000177494</td>\n",
       "      <td>ENSG00000177494</td>\n",
       "      <td>3</td>\n",
       "      <td>111314290</td>\n",
       "      <td>1.183398</td>\n",
       "      <td>0.892619</td>\n",
       "      <td>0.917497</td>\n",
       "      <td>3.504894</td>\n",
       "      <td>1.168247</td>\n",
       "      <td>3.110953</td>\n",
       "      <td>...</td>\n",
       "      <td>1.936493</td>\n",
       "      <td>1.177972</td>\n",
       "      <td>2.673129</td>\n",
       "      <td>13.892711</td>\n",
       "      <td>2.763749</td>\n",
       "      <td>3.032246</td>\n",
       "      <td>4.558199</td>\n",
       "      <td>2.807851</td>\n",
       "      <td>2.021329</td>\n",
       "      <td>3.181096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23722 rows × 466 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TargetID      Gene_Symbol Chr      Coord    HG00096    HG00097  \\\n",
       "0      ENSG00000152931  ENSG00000152931   5   59783540   0.101858   0.078110   \n",
       "1      ENSG00000183696  ENSG00000183696   7   48128225   8.183805   5.686911   \n",
       "2      ENSG00000139269  ENSG00000139269  12   57846106   1.199910   1.573572   \n",
       "3      ENSG00000169129  ENSG00000169129  10  116164515   0.831940   0.069778   \n",
       "4      ENSG00000134602  ENSG00000134602   X  131157293  27.646422  24.395572   \n",
       "...                ...              ...  ..        ...        ...        ...   \n",
       "23717  ENSG00000235472  ENSG00000235472  13   29172970  31.582832  34.071123   \n",
       "23718  ENSG00000114423  ENSG00000114423   3  105588396  14.054749  14.477899   \n",
       "23719  ENSG00000243312  ENSG00000243312   4   87791344   1.112114   0.831797   \n",
       "23720  ENSG00000257337  ENSG00000257337  12   53448222   3.826396   6.045798   \n",
       "23721  ENSG00000177494  ENSG00000177494   3  111314290   1.183398   0.892619   \n",
       "\n",
       "         HG00099    HG00100    HG00101    HG00102  ...    NA20810    NA20811  \\\n",
       "0       0.048981   0.118597   0.004035   0.010925  ...   0.088601   0.240010   \n",
       "1       2.434653   3.830894   6.612288   4.709646  ...  13.428205   6.094500   \n",
       "2       0.521616   1.447225   3.565791   1.982681  ...   3.225880   1.996067   \n",
       "3       0.931086   0.620941   1.660668   0.570481  ...   1.023381   1.127852   \n",
       "4      16.445374  24.806650  25.113349  19.233988  ...  25.079490  28.725528   \n",
       "...          ...        ...        ...        ...  ...        ...        ...   \n",
       "23717  19.394365  37.523721  33.430473  37.844966  ...  47.905182  33.224977   \n",
       "23718  11.584425  12.637956  12.015089  13.750655  ...  11.723462   9.900372   \n",
       "23719   0.253228   0.271568   0.486086   1.362640  ...   1.168991   0.645389   \n",
       "23720   2.593872   4.447169   5.294657   4.106823  ...   6.524398   6.017795   \n",
       "23721   0.917497   3.504894   1.168247   3.110953  ...   1.936493   1.177972   \n",
       "\n",
       "         NA20812    NA20813    NA20814    NA20815    NA20816    NA20819  \\\n",
       "0       0.137175   0.148494   0.038643   0.088509   0.029204   0.024423   \n",
       "1      12.536000   2.217262   3.573394   7.583364   4.052882   1.570378   \n",
       "2       2.854923   2.267343   1.331201   2.187895   1.004250   3.003316   \n",
       "3       0.774409   1.495854   0.895342   1.513521   0.826377   1.021201   \n",
       "4      24.450520  27.264069  26.912814  29.509210  26.462331  25.624009   \n",
       "...          ...        ...        ...        ...        ...        ...   \n",
       "23717  39.827675  27.096811  34.686342  37.386766  30.598840  33.516674   \n",
       "23718  10.473115  13.433413  15.832594  19.216176  10.213739  14.563192   \n",
       "23719   0.819469   0.515448   0.463054   1.580658   0.701396   0.771233   \n",
       "23720   3.460273   3.905032   5.023161   5.333027   5.694370   8.142939   \n",
       "23721   2.673129  13.892711   2.763749   3.032246   4.558199   2.807851   \n",
       "\n",
       "         NA20826    NA20828  \n",
       "0       0.044816   0.139186  \n",
       "1       4.900372   6.737308  \n",
       "2       1.984362   1.684954  \n",
       "3       0.952502   0.740565  \n",
       "4      25.707741  22.824957  \n",
       "...          ...        ...  \n",
       "23717  32.017940  38.341888  \n",
       "23718  15.637732   8.357117  \n",
       "23719   0.857330   0.703369  \n",
       "23720   5.622043   6.026476  \n",
       "23721   2.021329   3.181096  \n",
       "\n",
       "[23722 rows x 466 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_expression_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "328317ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6402f9fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotation_file = data_folder + 'gene_annot.txt'\n",
    "gene_annotation_data = pd.read_csv(gene_annotation_file, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "70a4b532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>STOP</th>\n",
       "      <th>SYM</th>\n",
       "      <th>TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11868</td>\n",
       "      <td>13052</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM87B</td>\n",
       "      <td>1</td>\n",
       "      <td>817370</td>\n",
       "      <td>819834</td>\n",
       "      <td>ENSG00000177757</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LINC00115</td>\n",
       "      <td>1</td>\n",
       "      <td>826205</td>\n",
       "      <td>827522</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01128</td>\n",
       "      <td>1</td>\n",
       "      <td>827607</td>\n",
       "      <td>859446</td>\n",
       "      <td>ENSG00000228794</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20495</th>\n",
       "      <td>ARSA</td>\n",
       "      <td>22</td>\n",
       "      <td>50622753</td>\n",
       "      <td>50628179</td>\n",
       "      <td>ENSG00000100299</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20496</th>\n",
       "      <td>SHANK3</td>\n",
       "      <td>22</td>\n",
       "      <td>50674414</td>\n",
       "      <td>50733298</td>\n",
       "      <td>ENSG00000251322</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20497</th>\n",
       "      <td>ACR</td>\n",
       "      <td>22</td>\n",
       "      <td>50738195</td>\n",
       "      <td>50745334</td>\n",
       "      <td>ENSG00000100312</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20498</th>\n",
       "      <td>RPL23AP82</td>\n",
       "      <td>22</td>\n",
       "      <td>50756947</td>\n",
       "      <td>50801309</td>\n",
       "      <td>ENSG00000184319</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20499</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>22</td>\n",
       "      <td>50767500</td>\n",
       "      <td>50783663</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20500 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID  CHR     START      STOP              SYM  \\\n",
       "0        DDX11L1    1     11868     13052  ENSG00000223972   \n",
       "1          OR4F5    1     69090     70008  ENSG00000186092   \n",
       "2         FAM87B    1    817370    819834  ENSG00000177757   \n",
       "3      LINC00115    1    826205    827522  ENSG00000225880   \n",
       "4      LINC01128    1    827607    859446  ENSG00000228794   \n",
       "...          ...  ...       ...       ...              ...   \n",
       "20495       ARSA   22  50622753  50628179  ENSG00000100299   \n",
       "20496     SHANK3   22  50674414  50733298  ENSG00000251322   \n",
       "20497        ACR   22  50738195  50745334  ENSG00000100312   \n",
       "20498  RPL23AP82   22  50756947  50801309  ENSG00000184319   \n",
       "20499     RABL2B   22  50767500  50783663  ENSG00000079974   \n",
       "\n",
       "                                     TYPE  \n",
       "0      transcribed_unprocessed_pseudogene  \n",
       "1                          protein_coding  \n",
       "2                                 lincRNA  \n",
       "3                                 lincRNA  \n",
       "4                    processed_transcript  \n",
       "...                                   ...  \n",
       "20495                      protein_coding  \n",
       "20496                      protein_coding  \n",
       "20497                      protein_coding  \n",
       "20498  transcribed_unprocessed_pseudogene  \n",
       "20499                      protein_coding  \n",
       "\n",
       "[20500 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972c81bb",
   "metadata": {},
   "source": [
    "### Subset gene_expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "98506d05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for protein-coding genes\n",
    "protein_coding_genes = gene_annotation_data[gene_annotation_data['TYPE'] == 'protein_coding']\n",
    "\n",
    "# Get a list of protein-coding gene symbols\n",
    "protein_coding_symbols = protein_coding_genes['SYM'].tolist()\n",
    "\n",
    "# Subset gene_expression_data for protein-coding genes\n",
    "protein_coding_expression_data = gene_expression_data[gene_expression_data['Gene_Symbol'].isin(protein_coding_symbols)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1bdf4a61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TargetID</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Coord</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>...</th>\n",
       "      <th>NA20810</th>\n",
       "      <th>NA20811</th>\n",
       "      <th>NA20812</th>\n",
       "      <th>NA20813</th>\n",
       "      <th>NA20814</th>\n",
       "      <th>NA20815</th>\n",
       "      <th>NA20816</th>\n",
       "      <th>NA20819</th>\n",
       "      <th>NA20826</th>\n",
       "      <th>NA20828</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000183696</td>\n",
       "      <td>ENSG00000183696</td>\n",
       "      <td>7</td>\n",
       "      <td>48128225</td>\n",
       "      <td>8.183805</td>\n",
       "      <td>5.686911</td>\n",
       "      <td>2.434653</td>\n",
       "      <td>3.830894</td>\n",
       "      <td>6.612288</td>\n",
       "      <td>4.709646</td>\n",
       "      <td>...</td>\n",
       "      <td>13.428205</td>\n",
       "      <td>6.094500</td>\n",
       "      <td>12.536000</td>\n",
       "      <td>2.217262</td>\n",
       "      <td>3.573394</td>\n",
       "      <td>7.583364</td>\n",
       "      <td>4.052882</td>\n",
       "      <td>1.570378</td>\n",
       "      <td>4.900372</td>\n",
       "      <td>6.737308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000139269</td>\n",
       "      <td>ENSG00000139269</td>\n",
       "      <td>12</td>\n",
       "      <td>57846106</td>\n",
       "      <td>1.199910</td>\n",
       "      <td>1.573572</td>\n",
       "      <td>0.521616</td>\n",
       "      <td>1.447225</td>\n",
       "      <td>3.565791</td>\n",
       "      <td>1.982681</td>\n",
       "      <td>...</td>\n",
       "      <td>3.225880</td>\n",
       "      <td>1.996067</td>\n",
       "      <td>2.854923</td>\n",
       "      <td>2.267343</td>\n",
       "      <td>1.331201</td>\n",
       "      <td>2.187895</td>\n",
       "      <td>1.004250</td>\n",
       "      <td>3.003316</td>\n",
       "      <td>1.984362</td>\n",
       "      <td>1.684954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000169129</td>\n",
       "      <td>ENSG00000169129</td>\n",
       "      <td>10</td>\n",
       "      <td>116164515</td>\n",
       "      <td>0.831940</td>\n",
       "      <td>0.069778</td>\n",
       "      <td>0.931086</td>\n",
       "      <td>0.620941</td>\n",
       "      <td>1.660668</td>\n",
       "      <td>0.570481</td>\n",
       "      <td>...</td>\n",
       "      <td>1.023381</td>\n",
       "      <td>1.127852</td>\n",
       "      <td>0.774409</td>\n",
       "      <td>1.495854</td>\n",
       "      <td>0.895342</td>\n",
       "      <td>1.513521</td>\n",
       "      <td>0.826377</td>\n",
       "      <td>1.021201</td>\n",
       "      <td>0.952502</td>\n",
       "      <td>0.740565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000136237</td>\n",
       "      <td>ENSG00000136237</td>\n",
       "      <td>7</td>\n",
       "      <td>22396763</td>\n",
       "      <td>3.788503</td>\n",
       "      <td>2.050963</td>\n",
       "      <td>4.000313</td>\n",
       "      <td>3.271619</td>\n",
       "      <td>1.798216</td>\n",
       "      <td>1.516688</td>\n",
       "      <td>...</td>\n",
       "      <td>2.909393</td>\n",
       "      <td>1.921176</td>\n",
       "      <td>5.083873</td>\n",
       "      <td>2.866573</td>\n",
       "      <td>1.297788</td>\n",
       "      <td>2.888316</td>\n",
       "      <td>2.145022</td>\n",
       "      <td>3.557598</td>\n",
       "      <td>4.152063</td>\n",
       "      <td>1.216834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000146072</td>\n",
       "      <td>ENSG00000146072</td>\n",
       "      <td>6</td>\n",
       "      <td>47277641</td>\n",
       "      <td>2.567374</td>\n",
       "      <td>2.030769</td>\n",
       "      <td>0.927560</td>\n",
       "      <td>1.351749</td>\n",
       "      <td>2.316378</td>\n",
       "      <td>0.931071</td>\n",
       "      <td>...</td>\n",
       "      <td>0.761671</td>\n",
       "      <td>0.797139</td>\n",
       "      <td>0.328339</td>\n",
       "      <td>0.156504</td>\n",
       "      <td>1.118230</td>\n",
       "      <td>0.240833</td>\n",
       "      <td>1.022178</td>\n",
       "      <td>0.393550</td>\n",
       "      <td>0.007050</td>\n",
       "      <td>0.642450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23712</th>\n",
       "      <td>ENSG00000087095</td>\n",
       "      <td>ENSG00000087095</td>\n",
       "      <td>17</td>\n",
       "      <td>26369182</td>\n",
       "      <td>5.490282</td>\n",
       "      <td>5.265422</td>\n",
       "      <td>3.457973</td>\n",
       "      <td>5.075550</td>\n",
       "      <td>5.571885</td>\n",
       "      <td>5.262668</td>\n",
       "      <td>...</td>\n",
       "      <td>6.678943</td>\n",
       "      <td>4.996638</td>\n",
       "      <td>5.125125</td>\n",
       "      <td>5.383470</td>\n",
       "      <td>5.206981</td>\n",
       "      <td>5.643264</td>\n",
       "      <td>5.290116</td>\n",
       "      <td>5.911672</td>\n",
       "      <td>5.249228</td>\n",
       "      <td>5.347071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23714</th>\n",
       "      <td>ENSG00000162144</td>\n",
       "      <td>ENSG00000162144</td>\n",
       "      <td>11</td>\n",
       "      <td>61129771</td>\n",
       "      <td>33.050457</td>\n",
       "      <td>23.159734</td>\n",
       "      <td>16.050400</td>\n",
       "      <td>20.848751</td>\n",
       "      <td>32.245357</td>\n",
       "      <td>26.842346</td>\n",
       "      <td>...</td>\n",
       "      <td>22.962417</td>\n",
       "      <td>26.718492</td>\n",
       "      <td>17.096197</td>\n",
       "      <td>36.692807</td>\n",
       "      <td>32.645718</td>\n",
       "      <td>26.245967</td>\n",
       "      <td>25.040414</td>\n",
       "      <td>33.586088</td>\n",
       "      <td>27.436847</td>\n",
       "      <td>24.236840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23715</th>\n",
       "      <td>ENSG00000129473</td>\n",
       "      <td>ENSG00000129473</td>\n",
       "      <td>14</td>\n",
       "      <td>23767999</td>\n",
       "      <td>5.960860</td>\n",
       "      <td>10.834208</td>\n",
       "      <td>4.606371</td>\n",
       "      <td>5.274462</td>\n",
       "      <td>6.453945</td>\n",
       "      <td>11.956208</td>\n",
       "      <td>...</td>\n",
       "      <td>7.947619</td>\n",
       "      <td>5.826466</td>\n",
       "      <td>6.691473</td>\n",
       "      <td>11.671064</td>\n",
       "      <td>9.606188</td>\n",
       "      <td>7.016099</td>\n",
       "      <td>6.305954</td>\n",
       "      <td>6.498428</td>\n",
       "      <td>4.919154</td>\n",
       "      <td>7.285334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23718</th>\n",
       "      <td>ENSG00000114423</td>\n",
       "      <td>ENSG00000114423</td>\n",
       "      <td>3</td>\n",
       "      <td>105588396</td>\n",
       "      <td>14.054749</td>\n",
       "      <td>14.477899</td>\n",
       "      <td>11.584425</td>\n",
       "      <td>12.637956</td>\n",
       "      <td>12.015089</td>\n",
       "      <td>13.750655</td>\n",
       "      <td>...</td>\n",
       "      <td>11.723462</td>\n",
       "      <td>9.900372</td>\n",
       "      <td>10.473115</td>\n",
       "      <td>13.433413</td>\n",
       "      <td>15.832594</td>\n",
       "      <td>19.216176</td>\n",
       "      <td>10.213739</td>\n",
       "      <td>14.563192</td>\n",
       "      <td>15.637732</td>\n",
       "      <td>8.357117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23721</th>\n",
       "      <td>ENSG00000177494</td>\n",
       "      <td>ENSG00000177494</td>\n",
       "      <td>3</td>\n",
       "      <td>111314290</td>\n",
       "      <td>1.183398</td>\n",
       "      <td>0.892619</td>\n",
       "      <td>0.917497</td>\n",
       "      <td>3.504894</td>\n",
       "      <td>1.168247</td>\n",
       "      <td>3.110953</td>\n",
       "      <td>...</td>\n",
       "      <td>1.936493</td>\n",
       "      <td>1.177972</td>\n",
       "      <td>2.673129</td>\n",
       "      <td>13.892711</td>\n",
       "      <td>2.763749</td>\n",
       "      <td>3.032246</td>\n",
       "      <td>4.558199</td>\n",
       "      <td>2.807851</td>\n",
       "      <td>2.021329</td>\n",
       "      <td>3.181096</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13536 rows × 466 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TargetID      Gene_Symbol Chr      Coord    HG00096    HG00097  \\\n",
       "1      ENSG00000183696  ENSG00000183696   7   48128225   8.183805   5.686911   \n",
       "2      ENSG00000139269  ENSG00000139269  12   57846106   1.199910   1.573572   \n",
       "3      ENSG00000169129  ENSG00000169129  10  116164515   0.831940   0.069778   \n",
       "5      ENSG00000136237  ENSG00000136237   7   22396763   3.788503   2.050963   \n",
       "13     ENSG00000146072  ENSG00000146072   6   47277641   2.567374   2.030769   \n",
       "...                ...              ...  ..        ...        ...        ...   \n",
       "23712  ENSG00000087095  ENSG00000087095  17   26369182   5.490282   5.265422   \n",
       "23714  ENSG00000162144  ENSG00000162144  11   61129771  33.050457  23.159734   \n",
       "23715  ENSG00000129473  ENSG00000129473  14   23767999   5.960860  10.834208   \n",
       "23718  ENSG00000114423  ENSG00000114423   3  105588396  14.054749  14.477899   \n",
       "23721  ENSG00000177494  ENSG00000177494   3  111314290   1.183398   0.892619   \n",
       "\n",
       "         HG00099    HG00100    HG00101    HG00102  ...    NA20810    NA20811  \\\n",
       "1       2.434653   3.830894   6.612288   4.709646  ...  13.428205   6.094500   \n",
       "2       0.521616   1.447225   3.565791   1.982681  ...   3.225880   1.996067   \n",
       "3       0.931086   0.620941   1.660668   0.570481  ...   1.023381   1.127852   \n",
       "5       4.000313   3.271619   1.798216   1.516688  ...   2.909393   1.921176   \n",
       "13      0.927560   1.351749   2.316378   0.931071  ...   0.761671   0.797139   \n",
       "...          ...        ...        ...        ...  ...        ...        ...   \n",
       "23712   3.457973   5.075550   5.571885   5.262668  ...   6.678943   4.996638   \n",
       "23714  16.050400  20.848751  32.245357  26.842346  ...  22.962417  26.718492   \n",
       "23715   4.606371   5.274462   6.453945  11.956208  ...   7.947619   5.826466   \n",
       "23718  11.584425  12.637956  12.015089  13.750655  ...  11.723462   9.900372   \n",
       "23721   0.917497   3.504894   1.168247   3.110953  ...   1.936493   1.177972   \n",
       "\n",
       "         NA20812    NA20813    NA20814    NA20815    NA20816    NA20819  \\\n",
       "1      12.536000   2.217262   3.573394   7.583364   4.052882   1.570378   \n",
       "2       2.854923   2.267343   1.331201   2.187895   1.004250   3.003316   \n",
       "3       0.774409   1.495854   0.895342   1.513521   0.826377   1.021201   \n",
       "5       5.083873   2.866573   1.297788   2.888316   2.145022   3.557598   \n",
       "13      0.328339   0.156504   1.118230   0.240833   1.022178   0.393550   \n",
       "...          ...        ...        ...        ...        ...        ...   \n",
       "23712   5.125125   5.383470   5.206981   5.643264   5.290116   5.911672   \n",
       "23714  17.096197  36.692807  32.645718  26.245967  25.040414  33.586088   \n",
       "23715   6.691473  11.671064   9.606188   7.016099   6.305954   6.498428   \n",
       "23718  10.473115  13.433413  15.832594  19.216176  10.213739  14.563192   \n",
       "23721   2.673129  13.892711   2.763749   3.032246   4.558199   2.807851   \n",
       "\n",
       "         NA20826    NA20828  \n",
       "1       4.900372   6.737308  \n",
       "2       1.984362   1.684954  \n",
       "3       0.952502   0.740565  \n",
       "5       4.152063   1.216834  \n",
       "13      0.007050   0.642450  \n",
       "...          ...        ...  \n",
       "23712   5.249228   5.347071  \n",
       "23714  27.436847  24.236840  \n",
       "23715   4.919154   7.285334  \n",
       "23718  15.637732   8.357117  \n",
       "23721   2.021329   3.181096  \n",
       "\n",
       "[13536 rows x 466 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_coding_expression_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bd6acd7",
   "metadata": {},
   "source": [
    "## Subset SNPs within +-500 Kb windows, and then conduct eQTL analysis for each of the 22 chromosomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abe04250",
   "metadata": {},
   "source": [
    "#### Create a dataframe that will incorporate all results of 22 chromosomes, and create the function to read in genotype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a65c8824",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty DataFrame to store results\n",
    "all_results = pd.DataFrame()\n",
    "\n",
    "# Function to read PLINK files for a specific chromosome\n",
    "def read_plink_chromosome(chromosome_number):\n",
    "    plink_folder = data_folder + 'LDREF/'\n",
    "    plink_prefix = plink_folder + f'1000G.EUR.{chromosome_number}'\n",
    "    return pp.read_plink(plink_prefix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4609c5a8",
   "metadata": {},
   "source": [
    "#### Chromosome 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3cd1dd8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 11.63it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '1']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 1] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(1)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f81a64be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "344"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(intersection_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "adb1f289",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>STOP</th>\n",
       "      <th>SYM</th>\n",
       "      <th>TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11868</td>\n",
       "      <td>13052</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM87B</td>\n",
       "      <td>1</td>\n",
       "      <td>817370</td>\n",
       "      <td>819834</td>\n",
       "      <td>ENSG00000177757</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LINC00115</td>\n",
       "      <td>1</td>\n",
       "      <td>826205</td>\n",
       "      <td>827522</td>\n",
       "      <td>ENSG00000225880</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01128</td>\n",
       "      <td>1</td>\n",
       "      <td>827607</td>\n",
       "      <td>859446</td>\n",
       "      <td>ENSG00000228794</td>\n",
       "      <td>processed_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2139</th>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>1</td>\n",
       "      <td>248810448</td>\n",
       "      <td>248826633</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2140</th>\n",
       "      <td>MIR3124</td>\n",
       "      <td>1</td>\n",
       "      <td>248826376</td>\n",
       "      <td>248826443</td>\n",
       "      <td>ENSG00000264500</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2141</th>\n",
       "      <td>ZNF672</td>\n",
       "      <td>1</td>\n",
       "      <td>248838209</td>\n",
       "      <td>248849517</td>\n",
       "      <td>ENSG00000171161</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2142</th>\n",
       "      <td>ZNF692</td>\n",
       "      <td>1</td>\n",
       "      <td>248850005</td>\n",
       "      <td>248859085</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2143</th>\n",
       "      <td>PGBD2</td>\n",
       "      <td>1</td>\n",
       "      <td>248906195</td>\n",
       "      <td>248919946</td>\n",
       "      <td>ENSG00000185220</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2144 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID  CHR      START       STOP              SYM  \\\n",
       "0       DDX11L1    1      11868      13052  ENSG00000223972   \n",
       "1         OR4F5    1      69090      70008  ENSG00000186092   \n",
       "2        FAM87B    1     817370     819834  ENSG00000177757   \n",
       "3     LINC00115    1     826205     827522  ENSG00000225880   \n",
       "4     LINC01128    1     827607     859446  ENSG00000228794   \n",
       "...         ...  ...        ...        ...              ...   \n",
       "2139    SH3BP5L    1  248810448  248826633  ENSG00000175137   \n",
       "2140    MIR3124    1  248826376  248826443  ENSG00000264500   \n",
       "2141     ZNF672    1  248838209  248849517  ENSG00000171161   \n",
       "2142     ZNF692    1  248850005  248859085  ENSG00000171163   \n",
       "2143      PGBD2    1  248906195  248919946  ENSG00000185220   \n",
       "\n",
       "                                    TYPE  \n",
       "0     transcribed_unprocessed_pseudogene  \n",
       "1                         protein_coding  \n",
       "2                                lincRNA  \n",
       "3                                lincRNA  \n",
       "4                   processed_transcript  \n",
       "...                                  ...  \n",
       "2139                      protein_coding  \n",
       "2140                               miRNA  \n",
       "2141                      protein_coding  \n",
       "2142                      protein_coding  \n",
       "2143                      protein_coding  \n",
       "\n",
       "[2144 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "21bad4af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NA20581</th>\n",
       "      <th>HG00277</th>\n",
       "      <th>NA20799</th>\n",
       "      <th>HG00180</th>\n",
       "      <th>HG00264</th>\n",
       "      <th>HG00381</th>\n",
       "      <th>HG00372</th>\n",
       "      <th>NA20772</th>\n",
       "      <th>HG01791</th>\n",
       "      <th>HG00115</th>\n",
       "      <th>...</th>\n",
       "      <th>NA12718</th>\n",
       "      <th>HG00252</th>\n",
       "      <th>NA20504</th>\n",
       "      <th>NA20515</th>\n",
       "      <th>HG00326</th>\n",
       "      <th>NA07056</th>\n",
       "      <th>NA12778</th>\n",
       "      <th>HG00121</th>\n",
       "      <th>NA12413</th>\n",
       "      <th>NA12348</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2.749185</td>\n",
       "      <td>2.728544</td>\n",
       "      <td>2.730144</td>\n",
       "      <td>4.359051</td>\n",
       "      <td>4.651351</td>\n",
       "      <td>3.189750</td>\n",
       "      <td>2.993184</td>\n",
       "      <td>3.879844</td>\n",
       "      <td>3.191631</td>\n",
       "      <td>1.488884</td>\n",
       "      <td>...</td>\n",
       "      <td>3.803506</td>\n",
       "      <td>5.094178</td>\n",
       "      <td>5.915144</td>\n",
       "      <td>3.918478</td>\n",
       "      <td>4.542814</td>\n",
       "      <td>3.282750</td>\n",
       "      <td>4.319513</td>\n",
       "      <td>3.722620</td>\n",
       "      <td>7.392088</td>\n",
       "      <td>5.495256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>5.637124</td>\n",
       "      <td>5.055427</td>\n",
       "      <td>-1.095338</td>\n",
       "      <td>3.199341</td>\n",
       "      <td>6.612252</td>\n",
       "      <td>1.449575</td>\n",
       "      <td>6.634505</td>\n",
       "      <td>-1.951744</td>\n",
       "      <td>5.834815</td>\n",
       "      <td>1.847980</td>\n",
       "      <td>...</td>\n",
       "      <td>2.026301</td>\n",
       "      <td>1.578335</td>\n",
       "      <td>5.320268</td>\n",
       "      <td>0.809630</td>\n",
       "      <td>4.276337</td>\n",
       "      <td>13.203988</td>\n",
       "      <td>3.625861</td>\n",
       "      <td>3.861601</td>\n",
       "      <td>39.100838</td>\n",
       "      <td>13.060729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>0.066064</td>\n",
       "      <td>0.283074</td>\n",
       "      <td>0.069232</td>\n",
       "      <td>0.130111</td>\n",
       "      <td>0.059283</td>\n",
       "      <td>-0.012758</td>\n",
       "      <td>0.035123</td>\n",
       "      <td>0.034455</td>\n",
       "      <td>0.103233</td>\n",
       "      <td>0.065562</td>\n",
       "      <td>...</td>\n",
       "      <td>0.158268</td>\n",
       "      <td>0.042356</td>\n",
       "      <td>0.104186</td>\n",
       "      <td>0.026417</td>\n",
       "      <td>0.082395</td>\n",
       "      <td>0.062630</td>\n",
       "      <td>0.084719</td>\n",
       "      <td>0.116357</td>\n",
       "      <td>0.031215</td>\n",
       "      <td>0.047240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>0.543428</td>\n",
       "      <td>0.358011</td>\n",
       "      <td>1.359402</td>\n",
       "      <td>0.508256</td>\n",
       "      <td>0.663733</td>\n",
       "      <td>0.605640</td>\n",
       "      <td>0.743010</td>\n",
       "      <td>0.969807</td>\n",
       "      <td>0.512179</td>\n",
       "      <td>0.462170</td>\n",
       "      <td>...</td>\n",
       "      <td>0.357547</td>\n",
       "      <td>0.858662</td>\n",
       "      <td>0.615691</td>\n",
       "      <td>0.936637</td>\n",
       "      <td>0.592057</td>\n",
       "      <td>0.704004</td>\n",
       "      <td>0.766123</td>\n",
       "      <td>0.388523</td>\n",
       "      <td>0.424280</td>\n",
       "      <td>0.400706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>0.032682</td>\n",
       "      <td>0.018164</td>\n",
       "      <td>0.029108</td>\n",
       "      <td>0.134324</td>\n",
       "      <td>0.181505</td>\n",
       "      <td>0.271835</td>\n",
       "      <td>-0.031909</td>\n",
       "      <td>0.102757</td>\n",
       "      <td>0.243646</td>\n",
       "      <td>0.014957</td>\n",
       "      <td>...</td>\n",
       "      <td>0.167848</td>\n",
       "      <td>0.133654</td>\n",
       "      <td>0.088081</td>\n",
       "      <td>0.146101</td>\n",
       "      <td>-0.054541</td>\n",
       "      <td>0.041426</td>\n",
       "      <td>0.781877</td>\n",
       "      <td>-0.051825</td>\n",
       "      <td>0.135623</td>\n",
       "      <td>1.917622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23600</th>\n",
       "      <td>1.900319</td>\n",
       "      <td>1.264244</td>\n",
       "      <td>0.986276</td>\n",
       "      <td>1.128541</td>\n",
       "      <td>1.217299</td>\n",
       "      <td>1.741634</td>\n",
       "      <td>1.417147</td>\n",
       "      <td>1.002235</td>\n",
       "      <td>0.941493</td>\n",
       "      <td>1.232987</td>\n",
       "      <td>...</td>\n",
       "      <td>0.481206</td>\n",
       "      <td>1.886135</td>\n",
       "      <td>1.306398</td>\n",
       "      <td>3.003186</td>\n",
       "      <td>0.496197</td>\n",
       "      <td>2.355953</td>\n",
       "      <td>0.724095</td>\n",
       "      <td>2.141281</td>\n",
       "      <td>0.715529</td>\n",
       "      <td>0.688207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23616</th>\n",
       "      <td>4.065664</td>\n",
       "      <td>4.830504</td>\n",
       "      <td>4.265805</td>\n",
       "      <td>6.868198</td>\n",
       "      <td>10.231141</td>\n",
       "      <td>4.733259</td>\n",
       "      <td>3.483288</td>\n",
       "      <td>10.232945</td>\n",
       "      <td>5.956079</td>\n",
       "      <td>5.700984</td>\n",
       "      <td>...</td>\n",
       "      <td>4.020621</td>\n",
       "      <td>6.203718</td>\n",
       "      <td>4.813290</td>\n",
       "      <td>4.219717</td>\n",
       "      <td>5.116473</td>\n",
       "      <td>5.283924</td>\n",
       "      <td>4.951325</td>\n",
       "      <td>5.478567</td>\n",
       "      <td>9.554028</td>\n",
       "      <td>7.775267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23622</th>\n",
       "      <td>4.450475</td>\n",
       "      <td>2.900151</td>\n",
       "      <td>2.927159</td>\n",
       "      <td>5.171821</td>\n",
       "      <td>4.101903</td>\n",
       "      <td>1.642093</td>\n",
       "      <td>3.191596</td>\n",
       "      <td>5.731440</td>\n",
       "      <td>5.218274</td>\n",
       "      <td>2.506862</td>\n",
       "      <td>...</td>\n",
       "      <td>1.957616</td>\n",
       "      <td>4.795253</td>\n",
       "      <td>5.222848</td>\n",
       "      <td>6.665778</td>\n",
       "      <td>1.460992</td>\n",
       "      <td>7.212469</td>\n",
       "      <td>3.656479</td>\n",
       "      <td>2.855420</td>\n",
       "      <td>2.144132</td>\n",
       "      <td>2.601503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23680</th>\n",
       "      <td>2.251077</td>\n",
       "      <td>0.534504</td>\n",
       "      <td>0.938240</td>\n",
       "      <td>0.309582</td>\n",
       "      <td>2.451371</td>\n",
       "      <td>1.659281</td>\n",
       "      <td>1.196890</td>\n",
       "      <td>1.644671</td>\n",
       "      <td>1.694738</td>\n",
       "      <td>1.613490</td>\n",
       "      <td>...</td>\n",
       "      <td>1.682486</td>\n",
       "      <td>1.130446</td>\n",
       "      <td>-0.496618</td>\n",
       "      <td>1.085239</td>\n",
       "      <td>-0.080288</td>\n",
       "      <td>0.916997</td>\n",
       "      <td>0.843959</td>\n",
       "      <td>0.463381</td>\n",
       "      <td>1.662580</td>\n",
       "      <td>-0.142445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23702</th>\n",
       "      <td>0.669112</td>\n",
       "      <td>0.808429</td>\n",
       "      <td>1.361132</td>\n",
       "      <td>1.349903</td>\n",
       "      <td>0.538435</td>\n",
       "      <td>0.635392</td>\n",
       "      <td>1.207492</td>\n",
       "      <td>0.725004</td>\n",
       "      <td>1.095383</td>\n",
       "      <td>0.684383</td>\n",
       "      <td>...</td>\n",
       "      <td>1.480221</td>\n",
       "      <td>1.127916</td>\n",
       "      <td>0.537274</td>\n",
       "      <td>1.481794</td>\n",
       "      <td>1.445769</td>\n",
       "      <td>1.042087</td>\n",
       "      <td>0.989272</td>\n",
       "      <td>0.713843</td>\n",
       "      <td>0.440772</td>\n",
       "      <td>0.919568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1467 rows × 344 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        NA20581   HG00277   NA20799   HG00180    HG00264   HG00381   HG00372  \\\n",
       "14     2.749185  2.728544  2.730144  4.359051   4.651351  3.189750  2.993184   \n",
       "17     5.637124  5.055427 -1.095338  3.199341   6.612252  1.449575  6.634505   \n",
       "64     0.066064  0.283074  0.069232  0.130111   0.059283 -0.012758  0.035123   \n",
       "72     0.543428  0.358011  1.359402  0.508256   0.663733  0.605640  0.743010   \n",
       "74     0.032682  0.018164  0.029108  0.134324   0.181505  0.271835 -0.031909   \n",
       "...         ...       ...       ...       ...        ...       ...       ...   \n",
       "23600  1.900319  1.264244  0.986276  1.128541   1.217299  1.741634  1.417147   \n",
       "23616  4.065664  4.830504  4.265805  6.868198  10.231141  4.733259  3.483288   \n",
       "23622  4.450475  2.900151  2.927159  5.171821   4.101903  1.642093  3.191596   \n",
       "23680  2.251077  0.534504  0.938240  0.309582   2.451371  1.659281  1.196890   \n",
       "23702  0.669112  0.808429  1.361132  1.349903   0.538435  0.635392  1.207492   \n",
       "\n",
       "         NA20772   HG01791   HG00115  ...   NA12718   HG00252   NA20504  \\\n",
       "14      3.879844  3.191631  1.488884  ...  3.803506  5.094178  5.915144   \n",
       "17     -1.951744  5.834815  1.847980  ...  2.026301  1.578335  5.320268   \n",
       "64      0.034455  0.103233  0.065562  ...  0.158268  0.042356  0.104186   \n",
       "72      0.969807  0.512179  0.462170  ...  0.357547  0.858662  0.615691   \n",
       "74      0.102757  0.243646  0.014957  ...  0.167848  0.133654  0.088081   \n",
       "...          ...       ...       ...  ...       ...       ...       ...   \n",
       "23600   1.002235  0.941493  1.232987  ...  0.481206  1.886135  1.306398   \n",
       "23616  10.232945  5.956079  5.700984  ...  4.020621  6.203718  4.813290   \n",
       "23622   5.731440  5.218274  2.506862  ...  1.957616  4.795253  5.222848   \n",
       "23680   1.644671  1.694738  1.613490  ...  1.682486  1.130446 -0.496618   \n",
       "23702   0.725004  1.095383  0.684383  ...  1.480221  1.127916  0.537274   \n",
       "\n",
       "        NA20515   HG00326    NA07056   NA12778   HG00121    NA12413    NA12348  \n",
       "14     3.918478  4.542814   3.282750  4.319513  3.722620   7.392088   5.495256  \n",
       "17     0.809630  4.276337  13.203988  3.625861  3.861601  39.100838  13.060729  \n",
       "64     0.026417  0.082395   0.062630  0.084719  0.116357   0.031215   0.047240  \n",
       "72     0.936637  0.592057   0.704004  0.766123  0.388523   0.424280   0.400706  \n",
       "74     0.146101 -0.054541   0.041426  0.781877 -0.051825   0.135623   1.917622  \n",
       "...         ...       ...        ...       ...       ...        ...        ...  \n",
       "23600  3.003186  0.496197   2.355953  0.724095  2.141281   0.715529   0.688207  \n",
       "23616  4.219717  5.116473   5.283924  4.951325  5.478567   9.554028   7.775267  \n",
       "23622  6.665778  1.460992   7.212469  3.656479  2.855420   2.144132   2.601503  \n",
       "23680  1.085239 -0.080288   0.916997  0.843959  0.463381   1.662580  -0.142445  \n",
       "23702  1.481794  1.445769   1.042087  0.989272  0.713843   0.440772   0.919568  \n",
       "\n",
       "[1467 rows x 344 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_data_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8ea0f441",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd147a9",
   "metadata": {},
   "source": [
    "#### Chromosome 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "582f73c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  9.13it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '2']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 2] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(2)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cc8d6c1",
   "metadata": {},
   "source": [
    "#### Chromosome 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "28ff2426",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 10.32it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '3']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 3] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(3)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87a4abbf",
   "metadata": {},
   "source": [
    "#### Chromosome 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "cfba72e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  8.99it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '4']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 4] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(4)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e88c472",
   "metadata": {},
   "source": [
    "#### Chromosome 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a46171c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 14.60it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '5']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 5] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(5)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "797672a4",
   "metadata": {},
   "source": [
    "#### Chromosome 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1e353082",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00,  9.89it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '6']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 6] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(6)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d90f2f9a",
   "metadata": {},
   "source": [
    "#### Chromosome 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1ddfa4ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 13.69it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '7']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 7] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(7)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "845ee03f",
   "metadata": {},
   "source": [
    "#### Chromosome 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3c208bf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 13.37it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '8']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 8] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(8)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db94282d",
   "metadata": {},
   "source": [
    "#### Chromosome 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "130f46c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 22.89it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '9']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 9] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(9)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa20cb72",
   "metadata": {},
   "source": [
    "#### Chromosome 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5fa2bece",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 16.88it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '10']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 10] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(10)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "621de57d",
   "metadata": {},
   "source": [
    "#### Chromosome 11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "88ea37e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 11.12it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '11']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 11] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(11)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57f5ddb6",
   "metadata": {},
   "source": [
    "#### Chromosome 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c63f0cca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 15.73it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '12']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 12] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(12)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a89b852",
   "metadata": {},
   "source": [
    "#### Chromosome 13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "ae29c4a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 17.70it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '13']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 13] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(13)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5541b7",
   "metadata": {},
   "source": [
    "#### Chromosome 14"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f5795bb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 17.52it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '14']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 14] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(14)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7b52b06",
   "metadata": {},
   "source": [
    "#### Chromosome 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4e2cbcc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 18.15it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '15']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 15] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(15)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b08eacfd",
   "metadata": {},
   "source": [
    "#### Chromosome 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4401d70d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 15.88it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '16']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 16] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(16)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4851b7fc",
   "metadata": {},
   "source": [
    "#### Chromosome 17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3e53bf09",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 21.79it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '17']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 17] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(17)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed77e611",
   "metadata": {},
   "source": [
    "#### Chromosome 18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e4e8456b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 19.46it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '18']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 18] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(18)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28a42085",
   "metadata": {},
   "source": [
    "#### Chromosome 19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fcd6aa6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 24.30it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '19']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 19] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(19)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2be185e",
   "metadata": {},
   "source": [
    "#### Chromosome 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "43b07b17",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 28.58it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '20']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 20] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(20)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2cab4e6",
   "metadata": {},
   "source": [
    "#### Chromosome 21"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "23cb7d39",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 25.47it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '21']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 21] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(21)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad140aaf",
   "metadata": {},
   "source": [
    "#### Chromosome 22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ecf917eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 23.06it/s]\n"
     ]
    }
   ],
   "source": [
    "expression_data = protein_coding_expression_data[protein_coding_expression_data['Chr'] == '22']  \n",
    "gene_annot = gene_annotation_data[gene_annotation_data['CHR'] == 22] \n",
    "\n",
    "# Read plink files\n",
    "(bim, fam, bed) = read_plink_chromosome(22)\n",
    "\n",
    "#Find common individuals\n",
    "fam_ids = fam['fid'].tolist()\n",
    "expression_ids = expression_data.columns.tolist()\n",
    "intersection_ids = set(fam_ids).intersection(set(expression_ids))\n",
    "intersection_ids_list = list(intersection_ids)\n",
    "\n",
    "# Filter the expression data and genotype data to include only common individuals\n",
    "expression_data_filtered = expression_data[intersection_ids_list]\n",
    "bed_filtered = bed[:, fam[fam['fid'].isin(intersection_ids_list)].index].compute()\n",
    "\n",
    "#Cis-eQTL Analysis\n",
    "results = []\n",
    "window_size = 500000  # 500 Kb window size\n",
    "for gene_row in gene_annot.itertuples():\n",
    "    gene_id = gene_row.ID\n",
    "    gene_sym = gene_annot[gene_annot[\"ID\"] == gene_id][\"SYM\"].iloc[0]\n",
    "    gene_start = gene_row.START\n",
    "    gene_end = gene_row.STOP\n",
    "\n",
    "    # Filter SNPs within the window of the gene\n",
    "    cis_snps = bim[(bim['pos'] > gene_start - window_size) & (bim['pos'] < gene_end + window_size)]\n",
    "\n",
    "    if cis_snps.empty:\n",
    "        continue\n",
    "    gene_expression_df = expression_data_filtered[expression_data[\"TargetID\"] == gene_sym]\n",
    "\n",
    "    if gene_expression_df.empty:\n",
    "        continue\n",
    "\n",
    "    gene_expression = gene_expression_df.iloc[0]\n",
    "\n",
    "    for snp_row in cis_snps.itertuples():\n",
    "        snp_id = snp_row.snp\n",
    "        snp = bed_filtered[snp_row.Index, :]\n",
    "\n",
    "        if len(gene_expression) != len(snp):\n",
    "            # Handle mismatch in length between gene expression and SNP data\n",
    "            continue\n",
    "\n",
    "        # Add a constant term to the independent variables to represent the intercept\n",
    "        snp_with_intercept = sm.add_constant(snp)\n",
    "\n",
    "        # Perform linear regression\n",
    "        model = sm.OLS(gene_expression, snp_with_intercept)\n",
    "        result = model.fit()\n",
    "\n",
    "        # Check if there are enough parameters in the results\n",
    "        if len(result.params) < 2:\n",
    "            continue\n",
    "\n",
    "        pvalue = result.pvalues[1]\n",
    "        beta = result.params[1]\n",
    "        std_err = result.bse[1]\n",
    "\n",
    "        results.append({\n",
    "            'gene_id': gene_id,\n",
    "            'gene_symbol': gene_sym,\n",
    "            'snp_id': snp_id,\n",
    "            'pvalue': pvalue,\n",
    "            'OR': beta,\n",
    "            'SE': std_err,\n",
    "            'snp_position': snp_row.pos\n",
    "        })\n",
    "\n",
    "all_results = pd.concat([all_results] + [pd.DataFrame([res]) for res in results], ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44b3cce",
   "metadata": {},
   "source": [
    "### Total results of all 22 chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "39ee0d65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>snp_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3094315</td>\n",
       "      <td>0.298244</td>\n",
       "      <td>0.094348</td>\n",
       "      <td>0.090563</td>\n",
       "      <td>752566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>0.276524</td>\n",
       "      <td>0.098526</td>\n",
       "      <td>0.090400</td>\n",
       "      <td>752721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3131969</td>\n",
       "      <td>0.538234</td>\n",
       "      <td>0.058996</td>\n",
       "      <td>0.095756</td>\n",
       "      <td>754182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs1048488</td>\n",
       "      <td>0.317524</td>\n",
       "      <td>0.089695</td>\n",
       "      <td>0.089604</td>\n",
       "      <td>760912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3115850</td>\n",
       "      <td>0.302921</td>\n",
       "      <td>0.092290</td>\n",
       "      <td>0.089450</td>\n",
       "      <td>761147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183028</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs13056621</td>\n",
       "      <td>0.850607</td>\n",
       "      <td>-0.051107</td>\n",
       "      <td>0.271142</td>\n",
       "      <td>51181759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183029</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs3865766</td>\n",
       "      <td>0.322009</td>\n",
       "      <td>-0.178194</td>\n",
       "      <td>0.179672</td>\n",
       "      <td>51186228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183030</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs3888396</td>\n",
       "      <td>0.511601</td>\n",
       "      <td>0.164925</td>\n",
       "      <td>0.251014</td>\n",
       "      <td>51211392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183031</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs2238837</td>\n",
       "      <td>0.338360</td>\n",
       "      <td>-0.184742</td>\n",
       "      <td>0.192690</td>\n",
       "      <td>51212875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183032</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs28729663</td>\n",
       "      <td>0.726480</td>\n",
       "      <td>0.089485</td>\n",
       "      <td>0.255598</td>\n",
       "      <td>51219006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6183033 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id      gene_symbol      snp_id    pvalue        OR        SE  \\\n",
       "0        SAMD11  ENSG00000187634   rs3094315  0.298244  0.094348  0.090563   \n",
       "1        SAMD11  ENSG00000187634   rs3131972  0.276524  0.098526  0.090400   \n",
       "2        SAMD11  ENSG00000187634   rs3131969  0.538234  0.058996  0.095756   \n",
       "3        SAMD11  ENSG00000187634   rs1048488  0.317524  0.089695  0.089604   \n",
       "4        SAMD11  ENSG00000187634   rs3115850  0.302921  0.092290  0.089450   \n",
       "...         ...              ...         ...       ...       ...       ...   \n",
       "6183028  RABL2B  ENSG00000079974  rs13056621  0.850607 -0.051107  0.271142   \n",
       "6183029  RABL2B  ENSG00000079974   rs3865766  0.322009 -0.178194  0.179672   \n",
       "6183030  RABL2B  ENSG00000079974   rs3888396  0.511601  0.164925  0.251014   \n",
       "6183031  RABL2B  ENSG00000079974   rs2238837  0.338360 -0.184742  0.192690   \n",
       "6183032  RABL2B  ENSG00000079974  rs28729663  0.726480  0.089485  0.255598   \n",
       "\n",
       "         snp_position  \n",
       "0              752566  \n",
       "1              752721  \n",
       "2              754182  \n",
       "3              760912  \n",
       "4              761147  \n",
       "...               ...  \n",
       "6183028      51181759  \n",
       "6183029      51186228  \n",
       "6183030      51211392  \n",
       "6183031      51212875  \n",
       "6183032      51219006  \n",
       "\n",
       "[6183033 rows x 7 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dfb6384",
   "metadata": {},
   "source": [
    "### The following codes are used to store the all_results dataframe to a local spreadsheet, so that we don't have to run the above analysis everytime we restart the notebook (jump to \"Discovering the significant SNPs\" if you don't want to save it to local folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "fa377be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_rows = 1048576  # Excel's limit\n",
    "num_sections = len(all_results) // max_rows + 1\n",
    "\n",
    "# Create a Pandas Excel writer using XlsxWriter as the engine\n",
    "with pd.ExcelWriter('large_file.xlsx', engine='xlsxwriter') as writer:\n",
    "    for section in range(num_sections):\n",
    "        start_row = max_rows * section\n",
    "        end_row = start_row + max_rows\n",
    "        # Write each section of the DataFrame to a different sheet\n",
    "        all_results.iloc[start_row:end_row].to_excel(writer, sheet_name=f'Project1_eQTL_results{section + 1}', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7417ba01",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_pattern = '*.xlsx'  # Assumes files have .xlsx extension\n",
    "\n",
    "# Create a list of all Excel files in the directory\n",
    "excel_files = glob.glob(file_pattern)\n",
    "\n",
    "# List of sheet names to read\n",
    "sheet_names = ['Project1_eQTL_results1', 'Project1_eQTL_results2', 'Project1_eQTL_results3', \n",
    "               'Project1_eQTL_results4', 'Project1_eQTL_results5', 'Project1_eQTL_results6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d170c4dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['large_file.xlsx']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "excel_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e44a4fb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read each sheet in each Excel file and store in a list\n",
    "dfs = []\n",
    "for file in excel_files:\n",
    "    for sheet in sheet_names:\n",
    "        df = pd.read_excel(file, sheet_name=sheet)\n",
    "        dfs.append(df)\n",
    "saved_results = pd.concat(dfs, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "af2bf277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>snp_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3094315</td>\n",
       "      <td>0.298244</td>\n",
       "      <td>0.094348</td>\n",
       "      <td>0.090563</td>\n",
       "      <td>752566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3131972</td>\n",
       "      <td>0.276524</td>\n",
       "      <td>0.098526</td>\n",
       "      <td>0.090400</td>\n",
       "      <td>752721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3131969</td>\n",
       "      <td>0.538234</td>\n",
       "      <td>0.058996</td>\n",
       "      <td>0.095756</td>\n",
       "      <td>754182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs1048488</td>\n",
       "      <td>0.317524</td>\n",
       "      <td>0.089695</td>\n",
       "      <td>0.089604</td>\n",
       "      <td>760912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs3115850</td>\n",
       "      <td>0.302921</td>\n",
       "      <td>0.092290</td>\n",
       "      <td>0.089450</td>\n",
       "      <td>761147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183023</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs13056621</td>\n",
       "      <td>0.850607</td>\n",
       "      <td>-0.051107</td>\n",
       "      <td>0.271142</td>\n",
       "      <td>51181759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183024</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs3865766</td>\n",
       "      <td>0.322009</td>\n",
       "      <td>-0.178194</td>\n",
       "      <td>0.179672</td>\n",
       "      <td>51186228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183025</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs3888396</td>\n",
       "      <td>0.511601</td>\n",
       "      <td>0.164925</td>\n",
       "      <td>0.251014</td>\n",
       "      <td>51211392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183026</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs2238837</td>\n",
       "      <td>0.338360</td>\n",
       "      <td>-0.184742</td>\n",
       "      <td>0.192690</td>\n",
       "      <td>51212875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6183027</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs28729663</td>\n",
       "      <td>0.726480</td>\n",
       "      <td>0.089485</td>\n",
       "      <td>0.255598</td>\n",
       "      <td>51219006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6183028 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id      gene_symbol      snp_id    pvalue        OR        SE  \\\n",
       "0        SAMD11  ENSG00000187634   rs3094315  0.298244  0.094348  0.090563   \n",
       "1        SAMD11  ENSG00000187634   rs3131972  0.276524  0.098526  0.090400   \n",
       "2        SAMD11  ENSG00000187634   rs3131969  0.538234  0.058996  0.095756   \n",
       "3        SAMD11  ENSG00000187634   rs1048488  0.317524  0.089695  0.089604   \n",
       "4        SAMD11  ENSG00000187634   rs3115850  0.302921  0.092290  0.089450   \n",
       "...         ...              ...         ...       ...       ...       ...   \n",
       "6183023  RABL2B  ENSG00000079974  rs13056621  0.850607 -0.051107  0.271142   \n",
       "6183024  RABL2B  ENSG00000079974   rs3865766  0.322009 -0.178194  0.179672   \n",
       "6183025  RABL2B  ENSG00000079974   rs3888396  0.511601  0.164925  0.251014   \n",
       "6183026  RABL2B  ENSG00000079974   rs2238837  0.338360 -0.184742  0.192690   \n",
       "6183027  RABL2B  ENSG00000079974  rs28729663  0.726480  0.089485  0.255598   \n",
       "\n",
       "         snp_position  \n",
       "0              752566  \n",
       "1              752721  \n",
       "2              754182  \n",
       "3              760912  \n",
       "4              761147  \n",
       "...               ...  \n",
       "6183023      51181759  \n",
       "6183024      51186228  \n",
       "6183025      51211392  \n",
       "6183026      51212875  \n",
       "6183027      51219006  \n",
       "\n",
       "[6183028 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "saved_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "032f99e1",
   "metadata": {},
   "source": [
    "### Discovering the significant SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e180be0",
   "metadata": {},
   "source": [
    "#### Filter with P-Value of 0.01"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "316c1775",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>snp_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>rs4970421</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>-0.510259</td>\n",
       "      <td>0.162775</td>\n",
       "      <td>1108637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>rs9729550</td>\n",
       "      <td>0.003789</td>\n",
       "      <td>-0.841076</td>\n",
       "      <td>0.288513</td>\n",
       "      <td>1135242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>rs604618</td>\n",
       "      <td>0.007438</td>\n",
       "      <td>-1.257599</td>\n",
       "      <td>0.467054</td>\n",
       "      <td>1232319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>rs1739855</td>\n",
       "      <td>0.003529</td>\n",
       "      <td>-1.277167</td>\n",
       "      <td>0.434732</td>\n",
       "      <td>1233941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>rs2477782</td>\n",
       "      <td>0.002817</td>\n",
       "      <td>-1.300435</td>\n",
       "      <td>0.432199</td>\n",
       "      <td>1237036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182741</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs9628283</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>-2.913991</td>\n",
       "      <td>0.738166</td>\n",
       "      <td>50540766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182918</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs140524</td>\n",
       "      <td>0.009310</td>\n",
       "      <td>0.657288</td>\n",
       "      <td>0.251325</td>\n",
       "      <td>50960682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182932</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs6520149</td>\n",
       "      <td>0.001168</td>\n",
       "      <td>1.964061</td>\n",
       "      <td>0.599877</td>\n",
       "      <td>50990669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182933</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs6009921</td>\n",
       "      <td>0.001168</td>\n",
       "      <td>1.964061</td>\n",
       "      <td>0.599877</td>\n",
       "      <td>50991205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6182934</th>\n",
       "      <td>RABL2B</td>\n",
       "      <td>ENSG00000079974</td>\n",
       "      <td>rs6009995</td>\n",
       "      <td>0.001168</td>\n",
       "      <td>1.964061</td>\n",
       "      <td>0.599877</td>\n",
       "      <td>50994841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64163 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id      gene_symbol     snp_id    pvalue        OR        SE  \\\n",
       "95       SAMD11  ENSG00000187634  rs4970421  0.001869 -0.510259  0.162775   \n",
       "301       NOC2L  ENSG00000188976  rs9729550  0.003789 -0.841076  0.288513   \n",
       "342       NOC2L  ENSG00000188976   rs604618  0.007438 -1.257599  0.467054   \n",
       "343       NOC2L  ENSG00000188976  rs1739855  0.003529 -1.277167  0.434732   \n",
       "344       NOC2L  ENSG00000188976  rs2477782  0.002817 -1.300435  0.432199   \n",
       "...         ...              ...        ...       ...       ...       ...   \n",
       "6182741  RABL2B  ENSG00000079974  rs9628283  0.000096 -2.913991  0.738166   \n",
       "6182918  RABL2B  ENSG00000079974   rs140524  0.009310  0.657288  0.251325   \n",
       "6182932  RABL2B  ENSG00000079974  rs6520149  0.001168  1.964061  0.599877   \n",
       "6182933  RABL2B  ENSG00000079974  rs6009921  0.001168  1.964061  0.599877   \n",
       "6182934  RABL2B  ENSG00000079974  rs6009995  0.001168  1.964061  0.599877   \n",
       "\n",
       "         snp_position  \n",
       "95            1108637  \n",
       "301           1135242  \n",
       "342           1232319  \n",
       "343           1233941  \n",
       "344           1237036  \n",
       "...               ...  \n",
       "6182741      50540766  \n",
       "6182918      50960682  \n",
       "6182932      50990669  \n",
       "6182933      50991205  \n",
       "6182934      50994841  \n",
       "\n",
       "[64163 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = saved_results[saved_results[\"pvalue\"] < 0.01]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0bc078e",
   "metadata": {},
   "source": [
    "#### Sort the dataframe after filtration by ascending p-value and descending OR (effect sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e7cff723",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\13269\\AppData\\Local\\Temp\\ipykernel_14020\\1554700442.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['abs_OR'] = df['OR'].abs()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>OR</th>\n",
       "      <th>SE</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>abs_OR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>203400</th>\n",
       "      <td>STIL</td>\n",
       "      <td>ENSG00000123473</td>\n",
       "      <td>rs1105456</td>\n",
       "      <td>2.586935e-23</td>\n",
       "      <td>-25.804894</td>\n",
       "      <td>2.406922</td>\n",
       "      <td>47269984</td>\n",
       "      <td>25.804894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2363128</th>\n",
       "      <td>PPP1R17</td>\n",
       "      <td>ENSG00000106341</td>\n",
       "      <td>rs16875607</td>\n",
       "      <td>1.683112e-21</td>\n",
       "      <td>-0.573896</td>\n",
       "      <td>0.056271</td>\n",
       "      <td>31657503</td>\n",
       "      <td>0.573896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2363134</th>\n",
       "      <td>PPP1R17</td>\n",
       "      <td>ENSG00000106341</td>\n",
       "      <td>rs10262468</td>\n",
       "      <td>1.683112e-21</td>\n",
       "      <td>-0.573896</td>\n",
       "      <td>0.056271</td>\n",
       "      <td>31667929</td>\n",
       "      <td>0.573896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753984</th>\n",
       "      <td>CCDC85A</td>\n",
       "      <td>ENSG00000055813</td>\n",
       "      <td>rs12622380</td>\n",
       "      <td>4.921446e-20</td>\n",
       "      <td>-5.514587</td>\n",
       "      <td>0.564650</td>\n",
       "      <td>55702485</td>\n",
       "      <td>5.514587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4120391</th>\n",
       "      <td>CCNA1</td>\n",
       "      <td>ENSG00000133101</td>\n",
       "      <td>rs9546280</td>\n",
       "      <td>9.198354e-19</td>\n",
       "      <td>-3.228637</td>\n",
       "      <td>0.344107</td>\n",
       "      <td>36638466</td>\n",
       "      <td>3.228637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75974</th>\n",
       "      <td>MINOS1</td>\n",
       "      <td>ENSG00000173436</td>\n",
       "      <td>rs12565770</td>\n",
       "      <td>9.999321e-03</td>\n",
       "      <td>-3.986838</td>\n",
       "      <td>1.539138</td>\n",
       "      <td>19555060</td>\n",
       "      <td>3.986838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362430</th>\n",
       "      <td>GHRHR</td>\n",
       "      <td>ENSG00000106128</td>\n",
       "      <td>rs17159410</td>\n",
       "      <td>9.999359e-03</td>\n",
       "      <td>-0.023909</td>\n",
       "      <td>0.009230</td>\n",
       "      <td>30779725</td>\n",
       "      <td>0.023909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3271155</th>\n",
       "      <td>C10orf12</td>\n",
       "      <td>ENSG00000155640</td>\n",
       "      <td>rs943544</td>\n",
       "      <td>9.999718e-03</td>\n",
       "      <td>0.053292</td>\n",
       "      <td>0.020574</td>\n",
       "      <td>97148731</td>\n",
       "      <td>0.053292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5728081</th>\n",
       "      <td>ZNF444</td>\n",
       "      <td>ENSG00000167685</td>\n",
       "      <td>rs2278281</td>\n",
       "      <td>9.999788e-03</td>\n",
       "      <td>0.933497</td>\n",
       "      <td>0.360383</td>\n",
       "      <td>55668197</td>\n",
       "      <td>0.933497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29637</th>\n",
       "      <td>CAMTA1</td>\n",
       "      <td>ENSG00000171735</td>\n",
       "      <td>rs3810985</td>\n",
       "      <td>9.999945e-03</td>\n",
       "      <td>0.855700</td>\n",
       "      <td>0.330350</td>\n",
       "      <td>6484036</td>\n",
       "      <td>0.855700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64163 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id      gene_symbol      snp_id        pvalue         OR  \\\n",
       "203400       STIL  ENSG00000123473   rs1105456  2.586935e-23 -25.804894   \n",
       "2363128   PPP1R17  ENSG00000106341  rs16875607  1.683112e-21  -0.573896   \n",
       "2363134   PPP1R17  ENSG00000106341  rs10262468  1.683112e-21  -0.573896   \n",
       "753984    CCDC85A  ENSG00000055813  rs12622380  4.921446e-20  -5.514587   \n",
       "4120391     CCNA1  ENSG00000133101   rs9546280  9.198354e-19  -3.228637   \n",
       "...           ...              ...         ...           ...        ...   \n",
       "75974      MINOS1  ENSG00000173436  rs12565770  9.999321e-03  -3.986838   \n",
       "2362430     GHRHR  ENSG00000106128  rs17159410  9.999359e-03  -0.023909   \n",
       "3271155  C10orf12  ENSG00000155640    rs943544  9.999718e-03   0.053292   \n",
       "5728081    ZNF444  ENSG00000167685   rs2278281  9.999788e-03   0.933497   \n",
       "29637      CAMTA1  ENSG00000171735   rs3810985  9.999945e-03   0.855700   \n",
       "\n",
       "               SE  snp_position     abs_OR  \n",
       "203400   2.406922      47269984  25.804894  \n",
       "2363128  0.056271      31657503   0.573896  \n",
       "2363134  0.056271      31667929   0.573896  \n",
       "753984   0.564650      55702485   5.514587  \n",
       "4120391  0.344107      36638466   3.228637  \n",
       "...           ...           ...        ...  \n",
       "75974    1.539138      19555060   3.986838  \n",
       "2362430  0.009230      30779725   0.023909  \n",
       "3271155  0.020574      97148731   0.053292  \n",
       "5728081  0.360383      55668197   0.933497  \n",
       "29637    0.330350       6484036   0.855700  \n",
       "\n",
       "[64163 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['abs_OR'] = df['OR'].abs()\n",
    "\n",
    "# Sort the DataFrame first by pvalue in ascending order, then by abs_OR in descending order\n",
    "sorted_df = df.sort_values(by=['pvalue', 'abs_OR'], ascending=[True, False])\n",
    "\n",
    "sorted_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bded8536",
   "metadata": {},
   "source": [
    "#### The first row is the most significant SNP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3542143c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
